<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<!--  -->
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="application/xml+xhtml; charset=UTF-8"/>
<title>enclone (bit.ly/enclone)</title>
<link rel="stylesheet" type="text/css" href="pages/enclone.css">
</head>

                

                <! ––

                💩 💩 💩 🔴 🔨 🔨 🔨 🔨 🔨 🔨 🔴 💩 💩 💩

                PUT DOWN YOUR HAMMER.
                THIS IS AN AUTO-GENERATED FILE.  PLEASE DO NOT EDIT IT.
                THANK YOU FOR YOUR COOPERATION,

                SINCERELY,
                THE BENEVOLENT OVERLORDS

                💩 💩 💩 🔴 🔨 🔨 🔨 🔨 🔨 🔨 🔴 💩 💩 💩

                ––>

<body>

<br>
<img src="img/enclone_banner.png" alt="enclone banner" title="enclone banner" width=100% />

<p><b>enclone</b> (beta) is a computational tool for studying the adaptive immune system in 
humans and other vertebrate species.  It takes as input data from
<b><a href="https://10xgenomics.com/">10x Genomics</a></b>, providing captured RNA
sequences for B and T cell receptors in single cells.  It organizes these cells into groups
(clonotypes) arising from the same progenitor and compactly displays each clonotype along
with its salient features, including mutated amino acids.</p>

<p style="border: 2px; border-style: solid; border-color: black; background-color: #FFFDD0; 
padding: 8px; width: 875px; font-size: 120%">
enclone (beta) is provided as a tool for use by the community to accelerate immunology research.  enclone is only 
supported via 
<a href="mailto:enclone@10xgenomics.com">enclone@10xgenomics.com</a>.
The clonotype assignment algorithm that is part of enclone will be integrated into a future release of Cell Ranger. 
</p>

<p>enclone and this page are designed for immunologists, but anyone can download and experiment
with it.</p>
<p><b>Background:</b> when you get sick, your body mounts an immune response by selectively 
amplifying
immune cells and mutations within these selected cells.  enclone allows you to see the history of
single immune cells within a biological sample (such as a blood draw or biopsy).  This history
reflects how the cognate receptors of these cells evolved in response to antigens, including
viruses, bacteria, and tumors.</p>

<table rules="all" cellpadding="3" style="border: 1px solid black">
<tbody>

<tr>
<td align="left"><a href="#mission">&nbsp;1. The mission of enclone&nbsp;</a></td>
<td align="left"><a href="#output">&nbsp;8. Understanding enclone output&nbsp;</a></td>
</tr>

<tr>
<td align="left"><a href="#inputs">&nbsp;2. Inputs to enclone&nbsp;</a></td>
<td align="left"><a href="#multiomic">&nbsp;9. Combining multiomic data&nbsp;</a></td>
</tr>

<tr>
<td align="left"><a href="#why">&nbsp;3. Why use enclone&nbsp;</a></td>
<td align="left"><a href="#honeycomb">&nbsp;10. Visualizing many clonotypes at once&nbsp;</a></td>
</tr>

<tr>
<td align="left"><a href="#software">&nbsp;4. The enclone software&nbsp;</a></td>
<td align="left"><a href="#sowhat">&nbsp;11. So what is enclone good for?&nbsp;</a></td>
</tr>

<tr>
<td align="left"><a href="#download">&nbsp;5. Installing enclone&nbsp;</a></td>
<td align="left"><a href="#questions">&nbsp;12. Questions&nbsp;</a></td>
</tr>

<tr>
<td align="left"><a href="#running">&nbsp;6. Running enclone&nbsp;</a></td>
<td align="left"><a href="#whereami">&nbsp;13. Where am I?&nbsp;</a></td>
</tr>

<tr>
<td align="left"><a href="#help">&nbsp;7. Help me!&nbsp;</a></td>
<td align="left"></td>
</tr>

</tbody>
</table>

<br>

<hr>

<p><a name="mission" style="display:block; position:relative; top:-20px;"></a></p>
<h2>The mission of enclone</h2>

<p>If you have a sample and have generated single-cell VDJ data from the B or T cells within it, you
have the power to fully understand the nature of the receptors for those cells, because you have
in hand the sequences for each of their chains.</p>
<p>You should be able to directly see this biology, without aid from a computational expert.  To
that end, we start with this simple goal:</p>

<table style="width:100%; margin-top: -17px">
  <tr>
  <th>
       <p style="text-align: left; font-weight:normal; border: 2px; border-style: solid;
          padding: 8px; font-size: 120%">
       <span style="color:red">Find</span> and <span style="color:red">display</span> clonotypes:
       groups of T and B cells sharing the same fully rearranged common ancestor.
       </p>
  </th>
  <th style="padding-left: 20px; text-align: left; font-weight:normal">
       The <i>finding</i> part is
       algorithmically challenging: it is very easy to mistakenly put unrelated cells in the same
       clonotype, or to "pollute" a clonotype with chains that do not belong in it.
       The <i>displaying</i> part is also challenging, in part because it is hard to compactly 
       represent the wealth of available information.
  </th>
  </tr>
</table>

To make sure we are using terminology in the same way, the following diagram shows what a
<em>clonotype</em> is for B cells.  The same applies for T cells, but things are simpler because 
T cells do not have somatic hypermutation.

<img src="img/what_is_a_clonotype.svg" alt="what is a clonotype" title="what is a clonotype" style="max-width:80%;"></a>

<p>Each cell in a clonotype is typically represented by two or three chains.  Such information can
only be obtained from <em>single cell</em> data!  From such data, clonotypes can be computationally
approximated, with high accuracy (see below).  The method we use for this is described 
<a href="pages/auto/help.how.html">briefly in the online documentation for enclone</a>, 
and will be described separately in more detail.</p>

<p style="border: 2px; border-style: solid; border-color: black; padding: 8px;">
<b>Clonotyping performance.</b>  To test the performance of enclone, we combined data from 443 BCR 
libraries from 30 donors.  This yielded <span style="color:red">381506</span> clonotypes, of which
<span style="color:red">9573</span> contained at least two cells each.  Of these,
<span style="color:red">15 (0.16%)</span> contained data from multiple donors.  These multi-donor 
clonotypes are errors.  The fact that there are few of them suggests that the rate at which enclone
places unrelated cells in the same clonotype is low.  This is a consequence of both the paired 
chain data and the algorithm.
</p>

<hr>

<p><a name="inputs" style="display:block; position:relative; top:-20px;"></a></p>
<h2>Inputs to enclone</h2>

<p>
The minimal required inputs to enclone are single cell 5' data from 10x Genomics that read VDJ for 
single cells.  enclone can also make use of and display gene expression and feature barcode data 
from the same cells.  The latter can be used to quantify cell surface proteins, antigen binding, 
CRISPR guides, and other things.  You can see a list of publications that use 10x VDJ data
<a href="https://www.10xgenomics.com/resources/publications?sortBy=publications-relevance&refinementList%5BproductGroups%5D%5B0%5D=Single%20Cell%20Immune%20Profiling&page=1">here</a>.
</p>

<hr>

<p><a name="why" style="display:block; position:relative; top:-20px;"></a></p>
<h2>Why use enclone?</h2>
<p>
enclone has features you can't get elsewhere, and we are eager to hear your ideas for other
features and improvements.  enclone is also very fast, allowing analysis of datasets in seconds.  
You can install the software by typing one line and then see for yourself.  But most importantly, 
enclone provides the 10x Genomics perspective on your data and has been tested on more 10x Genomics
data than any other tool. We highly recommend using other tools as well, but we can guarantee you 
that in many cases, enclone will provide a different answer, which may affect the biological 
interpretation of your data.
</p>

<hr>

<p><a name="software" style="display:block; position:relative; top:-20px;"></a></p>
<h2>The enclone software</h2>

<p>enclone is beta software released under <a href="LICENSE.txt">this license</a>.  Binary 
executables for Linux and Mac can be
directly downloaded from this page, as can sample 10x Genomics datasets.  And then you're off and
running!  To use enclone, you need to know how to run command-line tools.  This is something that
can be learned easily, particularly if you have a friend or colleague who can help you
get started.  You do not need to be able to program, or anything of that sort.</p>
<p>enclone is fast, typically responding in seconds (if run on a single dataset).  It is intended
as an exploratory tool.  You can dynamically change your command line to select specific
clonotypes and fields you wish to see.  You can run enclone on a laptop or a desktop or a server.</p>
<p>enclone, in addition to 
<a href="https://support.10xgenomics.com/single-cell-vdj">Cell Ranger and Loupe</a>
(and in which the core algorithm of enclone will be integrated at a later point in time),
supports the analysis of VDJ and other data from the <a href="https://10xgenomics.com/solutions/vdj/">Chromium Single Cell Immune Profiling
solution</a>.</p>

<p>What does it mean that enclone is beta software?  It means that we are very actively 
developing it!  Features are still being shaped, and we may on rare occasions break syntax
that previously worked.  See <a href="pages/history.html">this page</a> for the history of
changes.</p>

<hr>

<p><a name="download" style="display:block; position:relative; top:-20px;"></a></p>
<h2>Installing enclone</h2>

<p>
You can run enclone directly from a <b>Linux</b> or <b>Mac</b> terminal window; see <a href="pages/windows.html">here</a> for <b>Windows</b> options; 
see <a href="pages/dang_i_cannot_install.html">here</a> if you have a problem.
</p>

<table>
<tr>
<th align=left style="font-weight: normal">Type this&nbsp;</th>
<th align=left style="padding: 3px; font-weight: normal; border: 2px solid black">
<pre style="margin-top:3px; margin-bottom:3px"><code>curl -sSf -L bit.ly/enclone_install | sh -s <span style="color:red; font-weight: bold">SIZE</span></code></pre>
</th>
<th align=left style="font-weight: normal">
&nbsp;where <code style="color:red; font-weight: bold">SIZE</code> is 
<code>small</code>, <code>medium</code> or <code>large</code>, according to:
</th>
</tr>
</table>

<br>

<table rules="all" cellpadding="3" style="border: 2px solid black">
  <tr>
  <th align=left style="font-weight: normal; border: 2px solid black">
       <p style="margin: 4px">
       <code>small</code>
       </p>
  </th>

  <th align=left style="font-weight: normal; border: 2px solid black">
       <p style="margin: 4px">
       load small dataset collection (one dataset, 123085)
       </p>
  </th>
  <th align=left style="font-weight: normal; border: 2px solid black">
       <p style="margin: 4px">
       30 MB
       </p>
  </th>
  <th align=left style="font-weight: normal; border: 2px solid black">
       <p style="margin: 4px">
       do this if your internet connection is sloooooooow as a 🐌
       </p>
  </th>
  </tr>
  <tr>
  <th align=left style="font-weight: normal; border: 2px solid black">
       <p style="margin: 4px">
       <code>medium</code>
       </p>
  </th>
  <th align=left style="font-weight: normal; border: 2px solid black">
       <p style="margin: 4px">
       load medium dataset collection
       </p>
  </th>
  <th align=left style="font-weight: normal; border: 2px solid black">
       <p style="margin: 4px">
       350 MB
       </p>
  </th>
  <th align=left style="font-weight: normal; border: 2px solid black">
       <p style="margin: 4px">
       do this if you're happy with a bunch of datasets
       </p>
  </th>
  </tr>
  <tr>
  <th align=left style="font-weight: normal; border: 2px solid black">
       <p style="margin: 4px">
       <code>large</code>
       </p>
  </th>
  <th align=left style="font-weight: normal; border: 2px solid black">
       <p style="margin: 4px">
       load large dataset collection
       </p>
  </th>
  <th align=left style="font-weight: normal; border: 2px solid black">
       <p style="margin: 4px">
       2600 MB
       </p>
  </th>
  <th align=left style="font-weight: normal; border: 2px solid black">
       <p style="margin: 4px">
       do this if you want to play with a huge pile of datasets
       </p>
  </th>
  </tr>
</table>

<p>The command does three things:</p>
<ol>
<li>Puts the enclone executable (for Linux or Mac as appropriate) in <code>~/bin</code>.</li>
<li>If needed, adds a line to your bash initialization file so that <code>~/bin</code> is included.
</li>
<li>Puts the enclone datasets in <code>~/enclone</code>.</li>
</ol>
And a bit more detail is <a href="pages/installation_details.html">here</a>.

Close and reopen your terminal window to refresh your path.  Then you're good to go!</p>

<p><span style="color:red; font-weight:bold">To update, type the same command!</span>
<span style="font-weight:bold">
Only needed files will be downloaded.  
See <a href="pages/history.html">history</a> for what has changed.</span></p>

<hr>

<p><a name="running" style="display:block; position:relative; top:-20px;"></a></p>
<h2>Running enclone</h2>

<p>Running enclone can be as simple as typing e.g.</p>
<pre><code>enclone BCR=/home/my_name/experiment_123
</code></pre>
<p>where the path is where your Cell Ranger outputs live, but there are many options to learn
about.  For example, if you want to combine many datasets, you can do that, but you probably
need to provide a metadata file that describes the datasets.  You can find most of the enclone
documentation within its online menus.  To get started you should:</p>
<ol>
<li>
<p>Type <code>enclone help</code>, to make sure your terminal window works for <code>enclone</code>.</p>
</li>
<li>
<p>Type <code>enclone</code> to get to the main enclone help menu.</p>
</li>
</ol>
<hr>

<p><a name="help" style="display:block; position:relative; top:-20px;"></a></p>
<h2>Help me!</h2>

<p>enclone help is split between material on this site and pages that are invocable from the enclone
command line.  All the latter pages are also listed here:</p>
<table rules="all" cellpadding="3" style="border: 1px solid black">
<thead>
<tr>
<th align="left">command</th>
<th align="left">what it provides</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><a href="pages/auto/help.setup.html">enclone help</a></td>
<td align="left">help to test for correct setup</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.main.html">enclone</a></td>
<td align="left">what you see here; guide to all the doc</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.quick.html">enclone help quick</a></td>
<td align="left">quick guide to getting started</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.how.html">enclone help how</a></td>
<td align="left">outline of how enclone works, see also heuristics page, below</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.command.html">enclone help command</a></td>
<td align="left">info about enclone command line argument processing</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.glossary.html">enclone help glossary</a></td>
<td align="left">glossary of terms used by enclone, and conventions</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.example1.html">enclone help example1</a></td>
<td align="left">explanation of an example</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.example2.html">enclone help example2</a></td>
<td align="left">example showing gene expression and feature barcodes</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.support.html">enclone help support</a></td>
<td align="left">how we can help</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.input.html">enclone help input</a></td>
<td align="left">how to provide input to enclone</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.input_tech.html">enclone help input_tech</a></td>
<td align="left">how to provide input to enclone (technical notes)</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.parseable.html">enclone help parseable</a></td>
<td align="left">parseable output</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.filter.html">enclone help filter</a></td>
<td align="left">clonotype filtering options, scanning for feature enrichment</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.special.html">enclone help special</a></td>
<td align="left">special filtering options</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.lvars.html">enclone help lvars</a></td>
<td align="left">lead column options</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.cvars.html">enclone help cvars</a></td>
<td align="left">per chain column options</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.amino.html">enclone help amino</a></td>
<td align="left">per chain column options for amino acids</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.display.html">enclone help display</a></td>
<td align="left">other clonotype display options</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.indels.html">enclone help indels</a></td>
<td align="left">insertion and deletion handling</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.color.html">enclone help color</a></td>
<td align="left">how enclone uses color, and related things</td>
<tr>
<td align="left"><a href="pages/auto/help.faq.html">enclone help faq</a></td>
<td align="left">frequently asked questions</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.developer.html">enclone help developer</a></td>
<td align="left">a few things for developers</td>
</tr>
<tr>
<td align="left"><a href="pages/auto/help.all.html">enclone help all</a></td>
<td align="left">concatenation of all the help pages (USE THIS TO SEARCH ALL THE HELP PAGES)</td>
</tr>
</tbody>
</table>

<br>

For completeness, here are all the other pages on this site:

<br><br>

<table rules="all" cellpadding="3" style="border: 1px solid black">
<thead>
<tr>
<th align="left">page</th>
<th align="left">audience</th>
</tr>
</thead>
<tbody>
<tr>
    <td align="left"><a href="pages/history.html">history of changes</a></td>
    <td align="left">everyone</td>
</tr>
<tr>
    <td align="left"><a href="pages/auto/expanded.html">detecting illusory clonotypes</a></td>
    <td align="left">everyone</td>
</tr>
<tr>
    <td align="left"><a href="pages/compile.html">how to compile enclone</a></td>
    <td align="left">people who want to contribute code</td>
</tr>
<tr>
    <td align="left"><a href="LICENSE.txt">license</a></td>
    <td align="left">everyone</td>
</tr>
<tr>
    <td align="left"><a href="pages/windows.html">Windows</a></td>
    <td align="left">people using Windows computers</td>
</tr>
<tr>
    <td align="left"><a href="pages/heuristics.html">notes on heuristics (only a little bit now)
         </a></td>
    <td align="left">people who want to know details on how the algorithm works</td>
</tr>
<tr>
    <td align="left"><a href="pages/plot.html">honeycomb plots</a></td>
    <td align="left">everyone</td>
</tr>
<tr>
    <td align="left"><a href="pages/auto/tree.html">making phylogenetic trees</a></td>
    <td align="left">everyone</td>
</tr>
<tr>
    <td align="left"><a href="pages/dang_i_cannot_install.html">installation troubleshooting
    </a></td> <td align="left">if you have trouble installing</td>
</tr>
<tr>
    <td align="left"><a href="pages/installation_details.html">installation details
    </a></td> <td align="left">if you're curious about what the install command does</td>
</tr>
<tr>
    <td align="left"><a href="pages/auto/innate.html">iNKT and MAIT cells
    </a></td> <td align="left">people interested in iNKT and MAIT cells</td>
</tr>
</tbody>
</table>

<br>

<hr>

<p><a name="output" style="display:block; position:relative; top:-20px;"></a></p>
<h2>Understanding enclone output</h2>

<p>The example below shows how enclone prints out clonotypes.  This is something you'll need
to study in order to use enclone successfully.  enclone comes with extensive online
documentation, and because you can easily play with the sample datasets, you can gradually
figure out how it all works.</p>

<p>
<img src="img/enclone_annotated_example.svg" alt="enclone annotated example" title="enclone annotated example" style="max-width:100%;"></a>
</p> 

<p>Notice the compression in two directions.  Vertically, rather than showing one line for every 
cell, we group cells into a single line if they have identical VDJ transcripts.  Horizontally, 
rather than showing all transcript positions, we only show "interesting" positions.  This is a 
flexible concept, and what we show by default are all positions exhibiting a difference from the 
reference and all positions in the CDR3.</p>
<p>The same exact output would be obtained by typing</p>
<pre><code>enclone BCR=123085 CDR3=CQQRSNWPPSITF
</code></pre>
<p>The directory <code>123085</code> is in the directory <code>~/enclone/datasets</code> and 
contains some files from a Cell Ranger run, obtained from a human ovarian cancer sample.</p>

<p style="border: 2px; border-style: solid; border-color: black; padding: 8px; width: 950px">
<b>How does enclone find my data?</b>
It uses a <i>search path</i> called <code>PRE</code> that is preset to 
<code>~/enclone/datasets,~/enclone/datasets2</code>, and which you can set to any value, either
by setting <code>PRE=...</code> on the command line, or by setting the environment variable
<code>ENCLONE_PRE</code>.  To find your data, enclone prepends <code>PRE</code> to the value of 
<code>BCR</code> or <code>TCR</code> given on the command line.
For example, all of the following argument combinations do the same thing:
<br>1. <code>BCR=123085</code> (using the default value of <code>PRE</code>)
<br>2. <code>PRE=~/enclone/datasets BCR=123085</code>
<br>3. <code>PRE=~/enclone BCR=datasets/123085</code>
<br>4. <code>BCR=~/enclone/datasets/123085</code>.
<br>There is also an argument <code>META</code> that is convenient for specifying multiple
datasets.  See <a href="pages/auto/help.input.html">here</a> for how.
</p>

<p>The argument <code>CDR3=CQQRSNWPPSITF</code> causes enclone to display only clonotypes in which 
the given CDR3 sequence occurs.  Many other filters are provided.  In the absence of filters, all 
clonotypes are shown.  Clonotypes are shown from largest to smallest, and the output is 
automatically paged, so you can scroll through it.</p>
<p>By default, enclone prints clonotypes in this human-readable form.  You can also instruct enclone
to print clonotypes in machine-readable forms that are suitable for input to other programs.</p>
<hr>

<p><a name="multiomic" style="display:block; position:relative; top:-20px;"></a></p>
<h2>Combining multiomic data</h2>

<p>Gene expression and feature barcode data can be displayed simultaneously with VDJ data.  For
example, here we add columns for the same clonotype, showing the median number of UMIs detected
for, all genes, a particular gene, and a particular antibody:</p>

<pre style='font-family: "DejaVuSansMono"; line-height: 110%'><span style="font-size: 14px"><span style="color:#4c7aff;font-weight:bold;">[1] GROUP = 1 CLONOTYPES = 54 CELLS</span>

[1.1] CLONOTYPE = 54 CELLS
┌───────────────────────────────────┬──────────────────────────────────────┬───────────────────────────────┐
│                                   │  <span style="font-weight:bold;">CHAIN 1</span>                             │  CHAIN 2                      │
│                                   │  <span style="font-weight:bold;">144.1.2|IGHV3-49 ◆ 53|IGHJ3</span>         │  282|IGKV3-11 ◆ 218|IGKJ5     │
│                                   ├──────────────────────────────────────┼───────────────────────────────┤
│                                   │   1 11111111111111111 1              │    1111111111111              │
│                                   │  51 11112222222222333 4              │  6 0001111111111              │
│                                   │  53 67890123456789012 1              │  4 7890123456789              │
│                                   │     ═══════CDR3══════                │    ═════CDR3════              │
│reference                          │  <span style="color:#72c0ff;font-weight:bold;">V</span><span style="color:#f37721;font-weight:bold;">V</span> ◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦<span style="color:#f37721;font-weight:bold;">W</span> <span style="color:#f37721;font-weight:bold;">S</span>              │  <span style="color:#f37721;font-weight:bold;">R</span> <span style="color:#e7bd20;">C</span><span style="color:#3e7dd1;font-weight:bold;">QQ</span>◦◦◦◦◦◦◦◦◦◦              │
│donor ref                          │  <span style="color:#f37721;font-weight:bold;">FV</span> ◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦<span style="color:#f37721;font-weight:bold;">W</span> <span style="color:#f37721;font-weight:bold;">S</span>              │  <span style="color:#f37721;font-weight:bold;">R</span> <span style="color:#e7bd20;">C</span><span style="color:#3e7dd1;font-weight:bold;">QQ</span>◦◦◦◦◦◦◦◦◦◦              │
├───────────────────────────────────┼──────────────────────────────────────┼───────────────────────────────┤
│<span style="font-weight:bold;">#   n    gex  IGHV3-49_g  CD19_ab  </span>│<span style="font-weight:bold;">  .x ................. x    u  const  </span>│<span style="font-weight:bold;">  x .x...........      u  const</span>│
│1  48   9743         301       72  │  <span style="color:#f37721;font-weight:bold;">FV</span> <span style="color:#e7bd20;">C</span><span style="color:#eb99c2;font-weight:bold;">T</span><span style="color:#e7bd20;">R</span><span style="color:#31c39a;">DR</span><span style="color:#3e7dd1;font-weight:bold;">D</span><span style="color:#31c39a;">L</span><span style="color:#f37721;font-weight:bold;">R</span><span style="color:#eb99c2;font-weight:bold;">GA</span><span style="color:#e7bd20;">T</span><span style="color:#3e7dd1;font-weight:bold;">D</span><span style="color:#72c0ff;font-weight:bold;">A</span><span style="color:#3e7dd1;font-weight:bold;">F</span><span style="color:#31c39a;">D</span><span style="color:#eb99c2;font-weight:bold;">I</span><span style="color:#f37721;font-weight:bold;">W</span> <span style="color:#f37721;font-weight:bold;">S</span>  101  IGHG1  │  <span style="color:#f37721;font-weight:bold;">R</span> <span style="color:#e7bd20;">C</span><span style="color:#3e7dd1;font-weight:bold;">QQ</span><span style="color:#72c0ff;font-weight:bold;">R</span><span style="color:#3e7dd1;font-weight:bold;">S</span><span style="color:#f37721;font-weight:bold;">NW</span><span style="color:#3e7dd1;font-weight:bold;">PP</span><span style="color:#72c0ff;font-weight:bold;">S</span><span style="color:#eb99c2;font-weight:bold;">I</span><span style="color:#e7bd20;">T</span><span style="color:#f37721;font-weight:bold;">F</span>   3454  IGKC │
│2   3   9224         544       53  │  <span style="color:#f37721;font-weight:bold;">F</span><span style="color:#31c39a;">M</span> <span style="color:#e7bd20;">C</span><span style="color:#eb99c2;font-weight:bold;">T</span><span style="color:#e7bd20;">R</span><span style="color:#31c39a;">DR</span><span style="color:#3e7dd1;font-weight:bold;">D</span><span style="color:#31c39a;">L</span><span style="color:#f37721;font-weight:bold;">R</span><span style="color:#eb99c2;font-weight:bold;">GA</span><span style="color:#e7bd20;">T</span><span style="color:#3e7dd1;font-weight:bold;">D</span><span style="color:#72c0ff;font-weight:bold;">A</span><span style="color:#3e7dd1;font-weight:bold;">F</span><span style="color:#31c39a;">D</span><span style="color:#eb99c2;font-weight:bold;">I</span><span style="color:#f37721;font-weight:bold;">W</span> <span style="color:#f37721;font-weight:bold;">S</span>   89  IGHG1  │  <span style="color:#f37721;font-weight:bold;">R</span> <span style="color:#e7bd20;">CH</span><span style="color:#3e7dd1;font-weight:bold;">Q</span><span style="color:#72c0ff;font-weight:bold;">R</span><span style="color:#3e7dd1;font-weight:bold;">S</span><span style="color:#f37721;font-weight:bold;">NW</span><span style="color:#3e7dd1;font-weight:bold;">PP</span><span style="color:#72c0ff;font-weight:bold;">S</span><span style="color:#eb99c2;font-weight:bold;">I</span><span style="color:#e7bd20;">T</span><span style="color:#f37721;font-weight:bold;">F</span>   7663  IGKC │
│3   1  15850        1530       78  │  <span style="color:#f37721;font-weight:bold;">FV</span> <span style="color:#e7bd20;">C</span><span style="color:#eb99c2;font-weight:bold;">T</span><span style="color:#e7bd20;">R</span><span style="color:#31c39a;">DR</span><span style="color:#3e7dd1;font-weight:bold;">D</span><span style="color:#31c39a;">L</span><span style="color:#f37721;font-weight:bold;">R</span><span style="color:#eb99c2;font-weight:bold;">GA</span><span style="color:#e7bd20;">T</span><span style="color:#3e7dd1;font-weight:bold;">D</span><span style="color:#72c0ff;font-weight:bold;">A</span><span style="color:#3e7dd1;font-weight:bold;">F</span><span style="color:#31c39a;">D</span><span style="color:#eb99c2;font-weight:bold;">I</span><span style="color:#f37721;font-weight:bold;">W</span> <span style="color:#f37721;font-weight:bold;">S</span>  172  IGHG1  │  <span style="color:#3e7dd1;font-weight:bold;">S</span> <span style="color:#e7bd20;">C</span><span style="color:#3e7dd1;font-weight:bold;">QQ</span><span style="color:#72c0ff;font-weight:bold;">R</span><span style="color:#3e7dd1;font-weight:bold;">S</span><span style="color:#f37721;font-weight:bold;">NW</span><span style="color:#3e7dd1;font-weight:bold;">PP</span><span style="color:#72c0ff;font-weight:bold;">S</span><span style="color:#eb99c2;font-weight:bold;">I</span><span style="color:#e7bd20;">T</span><span style="color:#f37721;font-weight:bold;">F</span>  12603  IGKC │
│4   1   1519          11       21  │  <span style="color:#f37721;font-weight:bold;">FV</span> <span style="color:#e7bd20;">C</span><span style="color:#eb99c2;font-weight:bold;">T</span><span style="color:#e7bd20;">R</span><span style="color:#31c39a;">DR</span><span style="color:#3e7dd1;font-weight:bold;">D</span><span style="color:#31c39a;">L</span><span style="color:#f37721;font-weight:bold;">R</span><span style="color:#eb99c2;font-weight:bold;">GA</span><span style="color:#e7bd20;">T</span><span style="color:#3e7dd1;font-weight:bold;">D</span><span style="color:#72c0ff;font-weight:bold;">A</span><span style="color:#3e7dd1;font-weight:bold;">F</span><span style="color:#31c39a;">D</span><span style="color:#eb99c2;font-weight:bold;">I</span><span style="color:#f37721;font-weight:bold;">W</span> <span style="color:#31c39a;">S</span>   33  IGHG1  │  <span style="color:#f37721;font-weight:bold;">R</span> <span style="color:#e7bd20;">C</span><span style="color:#3e7dd1;font-weight:bold;">QQ</span><span style="color:#72c0ff;font-weight:bold;">R</span><span style="color:#3e7dd1;font-weight:bold;">S</span><span style="color:#f37721;font-weight:bold;">NW</span><span style="color:#3e7dd1;font-weight:bold;">PP</span><span style="color:#72c0ff;font-weight:bold;">S</span><span style="color:#eb99c2;font-weight:bold;">I</span><span style="color:#e7bd20;">T</span><span style="color:#f37721;font-weight:bold;">F</span>    116  IGKC │
│5   1   5347         142       23  │                                      │  <span style="color:#f37721;font-weight:bold;">R</span> <span style="color:#e7bd20;">C</span><span style="color:#3e7dd1;font-weight:bold;">QQ</span><span style="color:#72c0ff;font-weight:bold;">R</span><span style="color:#3e7dd1;font-weight:bold;">S</span><span style="color:#f37721;font-weight:bold;">NW</span><span style="color:#3e7dd1;font-weight:bold;">PP</span><span style="color:#72c0ff;font-weight:bold;">S</span><span style="color:#eb99c2;font-weight:bold;">I</span><span style="color:#e7bd20;">T</span><span style="color:#f37721;font-weight:bold;">F</span>   3471  IGKC │
└───────────────────────────────────┴──────────────────────────────────────┴───────────────────────────────┘
</span></pre>

<p>To obtain this, we added the extra arguments
<code>GEX=123749 LVARSP=gex,IGHV3-49_g,CD19_ab</code>
to the previous command.  The <code>GEX</code> part points to the directory containing gene 
expression and feature barcode data.  The <code>LVARSP</code> part defines the additional columns 
to be displayed.</p>
<p>Other types of data can be brought in via featuring barcoding.  For example, response to
multiple antigens can be measured using 
<a href="https://www.ncbi.nlm.nih.gov/pubmed/31787378" rel="nofollow">LIBRA-seq</a>
and these data displayed as additional columns.</p>
<hr>

<p><a name="honeycomb" style="display:block; position:relative; top:-20px;"></a></p>
<h2>Visualizing many clonotypes at once</h2>

<table style="width:100%">
  <tr>
  <th>
       <img src="img/clono.svg" alt="honeycomb plot" title="honeycomb plot">
  </th>
  <th></th>
  <th></th>
  <th style="text-align: left; font-weight:normal">
      <p>You can select clonotypes in enclone and then display them using a "honeycomb" plot.</p>
      <p>In this instance, pre- and post-vaccination samples were collected from four individuals,
      many datasets were generated for each sample, and these were combined in a single call
      to enclone.  Clonotypes containing at least ten cells are shown.  
      The plot was generated by adding</p>
<pre><code>MIN_CELLS=10 PLOT="clono.svg,pre-&gt;blue,post-&gt;red</code>
<code>LEGEND=blue,"pre-vaccination cell",</code>
<code>       red,"post-vaccination cell"</code></pre>
      <p>to the enclone command line, yielding the image shown here as the file 
      <code>clono.svg</code>.</p>
      <p>For more information about honeycomb plots, see <a href="pages/plot.html">here</a>.</p>
  </th>
  </tr>
</table>

<br>
<hr>

<p><a name="sowhat" style="display:block; position:relative; top:-20px;"></a></p>
<h2>So what is enclone good for?</h2>

<p>There are many ways to use 10x data to study immune biology.  Thus in the previous section, the
red clonotypes may represent responses to antigens in the vaccine.  But for a given
disease <i>e.g.</i> COVID-19, one may not yet have a vaccine, and indeed the vaccine may be
the goal!  A different approach is needed.</p>
<p>One such approach is to <b> identify patient and survivor B cell
clonotypes that expand in response to infectious disease</b>. These define antibodies that can be
used to design passive or active vaccines.  Additional power is added by mapping
antigen specificity to multiple antigens directly via feature barcoding
(<a href="https://www.ncbi.nlm.nih.gov/pubmed/31787378" rel="nofollow">LIBRA-seq</a>).</p>
<p>These data are easy to display in enclone!  You can directly select candidates for
vaccine or therapeutic development by picking large clonotypes with high antigen counts and single
or multiple antigen specificities.</p>
<p>We are actively working on further functionality that will make this process even more effective.</p>

<p>See <a href="pages/auto/tree.html">this</a> for how to generate phylogenetic trees
using enclone.</p>

<p>Another example use of enclone shows how to <b><a href="pages/auto/expanded.html">detect illusory clonotypes</a></b>.</p>

<hr>

<p><a name="questions" style="display:block; position:relative; top:-20px;"></a></p>
<h2>Questions</h2>

<p>
Please ask us questions! We are greatly interested in your feedback and ideas you may have to
make enclone as useful as possible.
</p>

<p style="border: 2px; border-style: solid; border-color: black; background-color: #FFFDD0; 
padding: 8px; width: 875px; font-size: 120%">
<b>enclone</b> is provided as an tool for use by the community.  Although we cannot
guarantee full support for the software, please email us at 
<a href="mailto:enclone@10xgenomics.com">enclone@10xgenomics.com</a>
if you have problems, questions or comments 
(see <b><a href="pages/auto/help.support.html">enclone support</a></b>).
enclone is beta software and thus a work in progress.  We are actively making many changes and may 
be unable to respond promptly to your particular request.
</p>

<hr>

<p><a name="whereami" style="display:block; position:relative; top:-20px;"></a></p>
<h2>Where am I?</h2>

<!-- dunno why negative left margin needed to line things up on left: -->
<p style="margin-left: -1.5px; margin-top: 30px; font-family: DejaVuSansMono; font-size: 200%; color: blue">bit.ly/enclone</p>

</body>
</html>
