<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" 
"https://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<!--  -->
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="application/xml+xhtml; charset=UTF-8"/>
<title>enclone in cellranger</title>
<link rel="stylesheet" type="text/css" href="../enclone_css_v2.css">
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-58278925-3"></script>
<script>
window.dataLayer = window.dataLayer || [];
function gtag(){{dataLayer.push(arguments);}}
gtag('js', new Date());
gtag('config', 'UA-58278925-3');
</script>

        
        </head>

        <! â€“â€“

        ðŸ’© ðŸ’© ðŸ’© ðŸ”´ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”´ ðŸ’© ðŸ’© ðŸ’©

        PUT DOWN YOUR HAMMER.
        THIS IS AN AUTO-GENERATED FILE.  PLEASE DO NOT EDIT IT.
        THANK YOU FOR YOUR COOPERATION,

        SINCERELY,
        THE BENEVOLENT OVERLORDS

        ðŸ’© ðŸ’© ðŸ’© ðŸ”´ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”´ ðŸ’© ðŸ’© ðŸ’©

        â€“â€“>

<body>

<br>
<a href="../../index.html#help">
<img src="../../img/enclone_banner.png" alt="enclone banner" title="enclone banner" width=100% />
</a>

<h1><span style="color:rgb(120,123,175);font-weight:900">enclone</span> in cellranger</h1>

<p>
This page does two things: 
<ol>
<li>Show how to use the versions of <span style="color:rgb(120,123,175);font-weight:900">enclone</span> that were used by previous versions 
of Cell Ranger.</li>
<li>Show the <span style="color:rgb(120,123,175);font-weight:900">enclone</span> changes that affect Cell Ranger output.</li>
</ol>
</p>

<hr>

<h2>previous versions</h2>

<p>Previous releases of <span style="color:rgb(120,123,175);font-weight:900">enclone</span>, matching particular Cell Ranger releases, may be found here:
<br><a href="https://github.com/10XGenomics/enclone/releases/tag/v_cellranger_5.0.0">5.0.0</a>
(5.0.1 used the same <span style="color:rgb(120,123,175);font-weight:900">enclone</span> version)
<br><a href="https://github.com/10XGenomics/enclone/releases/tag/v_cellranger_6.0.0">6.0.0</a>
(6.0.1 used the same <span style="color:rgb(120,123,175);font-weight:900">enclone</span> version).
<br>To use these, click on the <code>enclone_linux</code> or <code>enclone_macos</code> link
(on one of those pages)
to download the binary executable file, then e.g. <code>chmod +x enclone_macos</code> to add
execute permission, then run with e.g. <code>enclone_macos</code>.
</p>

<hr>

<h2><span style="color:rgb(120,123,175);font-weight:900">enclone</span> changes that affect cellranger</h2>

<p>
<table rules="all" cellpadding="6" style="border: 1px solid black">
<thead>
<tr>
<th align="left"> date          </th>
<th align="left"> git hash      </th>
<th align="left"> first Cell Ranger version </th>
<th align="left"> change        </th>
</tr>
</thead>
<tbody>

<tr>
<td> Dec. 9, 2020 </td>
<td> <code>f50783</code> </td>
<td> 6.0 </td>
<td> 
<ul>
<li>Change the algorithm for deciding to use a donor reference allele, so that it checks all
    donor reference alleles for all V genes having the same name as the one originally assigned to
    a contig.</li>
<li>Change default <code>MAX_DIFFS</code> from 50 to 55.  Change default 
    <code>MAX_DEGRADATION</code> from 3 to 2.  Change default <code>MAX_SCORE</code> from 
    1,000,000 to 500,000.  Change default <code>MAX_CDR3_DIFFS</code> from 10 to 15.</li>
<li>Allow up to 5 (instead of 4) amino acid differences when joining chains.  This results in less 
    spurious chains in clonotypes.</li>
</ul>
</td>
</tr>

<tr>
<td> Dec. 9, 2020 </td>
<td> <code>1e1ef9</code> </td>
<td> 6.0 </td>
<td> 
<ul>
<li>Add a doublet test, which removes some exact subclonotypes that appear to represent doublets.
</li>
<li>Improve the disjoint orbit test.</li>
</ul>
</td>
</tr>

<tr>
<td> Feb. 13, 2021 </td>
<td> <code>a43c8c</code> </td>
<td> 6.1 </td>
<td> 
<ul>
<li> To determine assignment of chains to columns in a clonotype table, we now rely primarily on 
     the column information implied by the original joins of exact subclonotypes.  This eliminates 
     some artifactual behavior.</li>
<li> Change weak chain filtering to permit deletion of chains supported by up to 20 cells, rather
     than 5.  This fixes some pathological cases involving clonotypes having many chains.</li>
</ul>
</td>
</tr>

<tr>
<td> June 4, 2021 </td>
<td> <code>635832</code> </td>
<td> 6.1 </td>
<td> 
Tweak the order of default filters so as to increase the likelihood that clonotypes
that are held together tenuously (or not at all) will be pulled apart.  This has a tiny effect
on our datasets (changing about <code>1</code> in <code>10^5</code> clonotypes), but could affect
some user datasets disproportionally.
</td>
</tr>

<tr>
<td> Sept. 20, 2021 </td>
<td> <code>95c951</code> </td>
<td> TBD </td>
<td> 
Add a new default filter called <i>signature filtering</i> that under certain rare circumstances
will decompose complex clonotypes that are incorrectly glued together.  This filter can 
dramatically affect certain datasets, but has almost no effect on typical data (less than one 
per million clonotypes tested).
Please see <a href="../../pages/auto/default_filters.html">enclone default filters</a>.
</td>
</tr>

<tr>
<td> Dec. 13, 2021 </td>
<td> <code>803b05</code> </td>
<td> TBD </td>
<td> 
Make three changes to clonotyping that increase its specificity:
<ol>
<li>Reject joins where the number of differences in CDR3
is at least <code>CDR3_MULT</code> times the number of non-shared differences outside CDR3.  The
default value of <code>CDR3_MULT</code> is <code>5</code>.  This algorithmic change may be
overridden by using a large value instead, e.g. by <code>CDR3_MULT=100</code>.</li>
<li>Compute the number <code>N</code> in a different and
simpler fashion, as <code>80^cd</code>.  The old behavior may be obtained by adding the argument
<code>OLD_MULT</code>.</li>
<li>Prevent merger in cases where light chain constant
regions are different.  The old behavior maybe be obtained by adding the argument
<code>OLD_LIGHT</code>.</li>
</ol>
</td>
</tr>

<tr>
<td> Jan. 5, 2022 </td>
<td> <code>******</code> </td>
<td> TBD </td>
<td> 
Make five changes to BCR clonotyping that increase its sensitivity and specificity:
<ol>
<li>Increase the precision of the calculation that computes
the probability <code>p</code> that shared mutations occur by chance.  This was returning the
wrong answer in some cases.  The old behavior is no longer available.</li>
<li>Remove the cap on total differences between two exact subclonotypes in order for them to be
joined.  The old behavior may now be obtained by adding the argument <code>MAX_DIFFS=55</code>.
</li>
<li>Rather than impose a bound of <code>15</code> on the maximum number of nucleotide differences 
between CDR3 sequences, we now require at least <code>80%</code> identity (out of the total CDR3 
length).  For both of these, the measurement first computes the total number of differences, taking 
into account one heavy and one light chain.  The number <code>80</code> can be set to an alternate 
value using <code>JOIN_CDR3_IDENT=...</code>.  The old behavior can be obtained by setting
<code>JOIN_CDR3_IDENT=0</code> and <code>MAX_CDR3_DIFFS=15</code>.</li>
<li>Now compute the number <code>N</code> in a manner that takes into account CDR3 length.  To
do this, we first set <code>cd1</code> to the number of heavy chain CDR3 nucleotide differences,
and <code>cd2</code> to the number of light chain CDR3 nucleotide differences.  Let <code>n1</code>
be the nucleotide length of the heavy chain CDR3, and likewise <code>n2</code> for the light
chain.  Then <code>N</code> is set to 
<br>
<code>80^(42 * (cd1/n1 + cd2/n2))</code>.  The number 
<code>42</code> can be changed using the variable <code>CDR3_NORMAL_LEN</code>.  The behavior 
preceding Dec. 13 is still available by via the argument <code>OLD_MULT</code>.</li>
<li>The light chain constant region identity requirement is now relaxed to apply only when
<code>cd = 0</code>.  The relaxation cannot be turned off but the light chain requirement can
still be turned off entirely using <code>OLD_LIGHT</code>.</li>
</ol>
</td>
</tr>

</tbody>
</table>
</p>

</body>
</html>
