<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<!--  -->
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="application/xml+xhtml; charset=UTF-8"/>
<title>D gene example2</title>
<style type="text/css">
@font-face {
    font-family: "DejaVuSansMono";
    src:
    local("Menlo Regular"),
    url('https://cdn.jsdelivr.net/npm/@deathbeds/jupyterlab-font-dejavu-sans-mono@1.0.0/style/fonts/DejaVuSansMono.woff2')
    format("woff2"),
    url('https://cdn.jsdelivr.net/npm/dejavu-fonts-ttf@2.37.3/ttf/DejaVuSansMono.ttf')
    format("truetype"),
    url('../fonts/DejaVuLGCSansMono.ttf') format("truetype"),
    local("Courier New");
}
@font-face {
    font-family: "DejaVuSansMono";
    font-weight: bold;
    src:
    local("Menlo Bold"),
    url('https://cdn.jsdelivr.net/npm/@deathbeds/jupyterlab-font-dejavu-sans-mono@1.0.0/style/fonts/DejaVuSansMono-Bold.woff2')
    format("woff2"),
    url('https://cdn.jsdelivr.net/npm/dejavu-fonts-ttf@2.37.3/ttf/DejaVuSansMono-Bold.ttf')
    format("truetype"),
    url('../fonts/DejaVuLGCSansMono-Bold.ttf') format("truetype"),
    local("Courier New Bold");
}
</style>
</head>
<body>
<pre style='font-family: "DejaVuSansMono"; line-height: 110%'><span style="font-size: 14px">SUMMARY STATISTICS
1. overall
   • number of datasets = 1
   • number of donors = 1
   • original number of cells = 1654
2. barcode fate
   ┌──────────┬────────────────────────────┐
   │barcodes  │  why deleted               │
   ├──────────┼────────────────────────────┤
   │     173  │  failed UMI filter         │
   │      54  │  failed IMPROPER filter    │
   │      25  │  failed FOURSIE_KILL filter│
   │      11  │  failed QUAL filter        │
   │       8  │  failed GRAPH_FILTER filter│
   │       5  │  failed WHITEF filter      │
   │       4  │  failed WEAK_CHAINS filter │
   │       1  │  failed DOUBLET filter     │
   │     281  │  total                     │
   └──────────┴────────────────────────────┘
3. for the selected clonotypes
   ┌────────┬────────────────────────┬──────────────────┬───────┐
   │chains  │  clonotypes with this  │  cells in these  │      %│
   │        │      number of chains  │      clonotypes  │       │
   ├────────┼────────────────────────┼──────────────────┼───────┤
   │1       │                   139  │             140  │   10.2│
   │2       │                   295  │             959  │   69.8│
   │3       │                    16  │             262  │   19.1│
   │4       │                     4  │              12  │    0.9│
   │total   │                   454  │            1373  │  100.0│
   └────────┴────────────────────────┴──────────────────┴───────┘
   • number of clonotypes having at least two cells = 144
   • number of intradonor cell-cell merges = 919
   • number of intradonor cell-cell merges (quadratic) = 25,915
   • number of cells having 1 chain = 218
   • number of cells having 2 or 3 chains = 1152
   • estimated B-B doublet rate = 0.3% = 3/942 = cells with 4 chains / cells with 2 or 4 chains
   • mean over middle third of contig UMI counts (heavy chain) = 331.32
   • mean over middle third of contig UMI counts (light chain) = 1860.48
   • mean over middle third of cell UMI counts for cells having two chains = 3141.60
   • mean UMIs per cell = 5845.16
   • mean UMIs per cell having two chains = 6276.57
   • for reads contributing to UMIs in reported chains, mean reads per UMI = 6.25

GLOBAL VARIABLES

d_inconsistent_% = 2.28
d_inconsistent_n = 747
</span></pre>
</body>
</html>

