
[01mper-chain column options[0m: These options define per-chain variables, which correspond to columns
that appear once for each chain in each clonotype, and have one entry for each exact subclonotype.

Per-column variables are specified using
[01mCVARS=x1,...,xn[0m
where each xi is one of:

â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚var         â”‚  bases at positions in chain that vary across the clonotype                   â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚u           â”‚  [31mâ—[0m VDJ UMI count for each exact subclonotype, median across cells             â”‚
â”‚r           â”‚  [31mâ—[0m VDJ read count for each exact subclonotype, median across cells            â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚const       â”‚  constant region name                                                         â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚edit        â”‚  a string that partially defines the edit of the reference V(D)J concatenationâ”‚
â”‚            â”‚  that gives rise to the observed CDR3; this uses a coordinate system in which â”‚
â”‚            â”‚  0 is the first base of the J ref segment (or the first base of the D ref     â”‚
â”‚            â”‚  segment for IGH and TRB); for example D-4:4 denotes the deletion of the last â”‚
â”‚            â”‚  4 bases of the V segment, I0:2 denotes an insertion of 2 bases after the V   â”‚
â”‚            â”‚  and I0:2;S5 denotes that plus a substitution at position 5; in computing     â”‚
â”‚            â”‚  "edit", for IGH and TRB, we always test every possible D segment,            â”‚
â”‚            â”‚  regardless of whether one is annotated, and pick the best one; for this      â”‚
â”‚            â”‚  reason, "edit" may be slow                                                   â”‚
â”‚comp        â”‚  a measure of CDR3 complexity, which is the total number of S, D and I        â”‚
â”‚            â”‚  symbols in "edit" as defined above                                           â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚cdr3_dna    â”‚  the CDR3_DNA sequence                                                        â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚vjlen       â”‚  number of bases from the start of the V region to the end of the J region    â”‚
â”‚clen        â”‚  length of observed constant region (usually truncated at primer start)       â”‚
â”‚ulen        â”‚  length of observed 5'-UTR sequence                                           â”‚
â”‚cdiff       â”‚  differences with universal reference constant region, shown in the           â”‚
â”‚            â”‚  abbreviated form e.g. 22T (ref changed to T at base 22) or 22T+10            â”‚
â”‚            â”‚  (same but contig has 10 additional bases beyond end of ref C region          â”‚
â”‚            â”‚  At most five differences are shown, and if there are more, ... is appended.  â”‚
â”‚udiff       â”‚  like cdiff, but for the 5'-UTR                                               â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚notes       â”‚  optional note if there is an insertion or the end of J does not exactly abut â”‚
â”‚            â”‚  the beginning of C; elided if empty; also single base overlaps between       â”‚
â”‚            â”‚  J and C are not shown unless you use the special option [01mJC1[0m; we do this      â”‚
â”‚            â”‚  because with some VDJ references, one nearly always has such an overlap      â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚ndiff<n>vj  â”‚  number of base differences within V..J between this exact subclonotype and   â”‚
â”‚            â”‚  exact subclonotype n                                                         â”‚
â”‚d_univ      â”‚  distance from universal reference, more specifically,                        â”‚
â”‚            â”‚  number of base differences within V..J between this exact                    â”‚
â”‚            â”‚  clonotype and universal reference, exclusive of indels, the last 15          â”‚
â”‚            â”‚  bases of the V and the first 15 bases of the J                               â”‚
â”‚d_donor     â”‚  distance from donor reference,                                               â”‚
â”‚            â”‚  as above but computed using donor reference                                  â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜

[47m [0m [31mâ—[0m These variables have some alternate versions, as shown in the table below:
[47m [0m 
[47m [0m â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
[47m [0m â”‚variable  â”‚  semantics                    â”‚  visual  â”‚  visual      â”‚  parseable  â”‚  parseable â”‚
[47m [0m â”‚          â”‚                               â”‚          â”‚  (one cell)  â”‚             â”‚  (one cell)â”‚
[47m [0m â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
[47m [0m â”‚x         â”‚  median over cells            â”‚  yes     â”‚  this cell   â”‚  yes        â”‚  yes       â”‚
[47m [0m â”‚x_mean    â”‚  mean over cells              â”‚  yes     â”‚  null        â”‚  yes        â”‚  yes       â”‚
[47m [0m â”‚x_Î¼       â”‚  (same as above)              â”‚  yes     â”‚  null        â”‚  yes        â”‚  yes       â”‚
[47m [0m â”‚x_sum     â”‚  sum over cells               â”‚  yes     â”‚  null        â”‚  yes        â”‚  yes       â”‚
[47m [0m â”‚x_Î£       â”‚  (same as above)              â”‚  yes     â”‚  null        â”‚  yes        â”‚  yes       â”‚
[47m [0m â”‚x_min     â”‚  min over cells               â”‚  yes     â”‚  null        â”‚  yes        â”‚  yes       â”‚
[47m [0m â”‚x_max     â”‚  max over cells               â”‚  yes     â”‚  null        â”‚  yes        â”‚  yes       â”‚
[47m [0m â”‚x_%       â”‚  % of total GEX (genes only)  â”‚  yes     â”‚  this cell   â”‚  yes        â”‚  yes       â”‚
[47m [0m â”‚x_cell    â”‚  this cell                    â”‚  no      â”‚  no          â”‚  no         â”‚  this cell â”‚
[47m [0m â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
[47m [0m Some explanation is required.  If you use enclone without certain options, you get the "visual"
[47m [0m column.
[47m [0m â€¢ Add the option [01mPER_CELL[0m (see "enclone help display") and then you get visual output with extra
[47m [0m lines for each cell within an exact subclonotype, and each of those extra lines is described by
[47m [0m the "visual (one cell)" column.
[47m [0m â€¢ If you generate parseable output (see "enclone help parseable"), then you get the "parseable"
[47m [0m column for that output, unless you specify [01mPCELL[0m, and then you get the last column.
[47m [0m â€¢ For the forms with Î¼ and Î£, the Greek letters are only used in column headings for visual output
[47m [0m (to save space), and optionally, in names of fields on the command line.
[47m [0m [32mâ–¶[0m If you try out these features, you'll see exactly what happens! [32mâ—€[0m

At least one variable must be listed.  The default is [01mu,const,notes[0m.  [01mCVARSP[0m: same as [01mCVARS[0m but
appends.

