<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<!--  -->
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="application/xml+xhtml; charset=UTF-8"/>
<title>enclone help cvars</title>
<style type="text/css">
@font-face {
    font-family: "DejaVuSansMono";
    src:
    local("Menlo Regular"),
    url('https://cdn.jsdelivr.net/npm/@deathbeds/jupyterlab-font-dejavu-sans-mono@1.0.0/style/fonts/DejaVuSansMono.woff2')
    format("woff2"),
    url('https://cdn.jsdelivr.net/npm/dejavu-fonts-ttf@2.37.3/ttf/DejaVuSansMono.ttf')
    format("truetype"),
    url('../fonts/DejaVuLGCSansMono.ttf') format("truetype"),
    local("Courier New");
}
@font-face {
    font-family: "DejaVuSansMono";
    font-weight: bold;
    src:
    local("Menlo Bold"),
    url('https://cdn.jsdelivr.net/npm/@deathbeds/jupyterlab-font-dejavu-sans-mono@1.0.0/style/fonts/DejaVuSansMono-Bold.woff2')
    format("woff2"),
    url('https://cdn.jsdelivr.net/npm/dejavu-fonts-ttf@2.37.3/ttf/DejaVuSansMono-Bold.ttf')
    format("truetype"),
    url('../fonts/DejaVuLGCSansMono-Bold.ttf') format("truetype"),
    local("Courier New Bold");
}
</style>
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-58278925-3"></script>
<script>
window.dataLayer = window.dataLayer || [];
function gtag(){{dataLayer.push(arguments);}}
gtag('js', new Date());
gtag('config', 'UA-58278925-3');
</script>
</head>
<body>

<br>
<img src="../../img/enclone_banner.png" alt="enclone banner" title="enclone banner" width=100% />
<pre style='font-family: "DejaVuSansMono"; line-height: 110%'><span style="font-size: 14px">
<span style="font-weight:bold;">per-chain column options</span>: These options define per-chain variables, which correspond to columns
that appear once for each chain in each clonotype, and have one entry for each exact subclonotype.

Per-column variables are specified using
<span style="font-weight:bold;">CVARS=x1,...,xn</span>
where each xi is one of:

┌────────────┬──────────────────────────────────────────────────────────────────────────────┐
│var         │  bases at positions in chain that vary across the clonotype                  │
├────────────┼──────────────────────────────────────────────────────────────────────────────┤
│u           │  <span style="color:#c23621;">●</span> VDJ UMI count for each exact subclonotype, median across cells            │
│r           │  <span style="color:#c23621;">●</span> VDJ read count for each exact subclonotype, median across cells           │
├────────────┼──────────────────────────────────────────────────────────────────────────────┤
│const       │  constant region name                                                        │
├────────────┼──────────────────────────────────────────────────────────────────────────────┤
│edit        │  a string that defines the edit of the reference V(D)J concatenation versus  │
│            │  the contig, from the beginning of the CDR3 to the end of the J segment;     │
│            │  this uses a coordinate system in which 0 is the first base of the J ref     │
│            │  segment (or the first base of the D ref segment for IGH and TRB); for       │
│            │  example D-4:4 denotes the deletion of the last 4 bases of the V segment,    │
│            │  I0:2 denotes an insertion of 2 bases after the V                            │
│            │  and I0:2•S5 denotes that plus a substitution at position 5; in computing    │
│            │  "edit", for IGH and TRB, we always test every possible D segment,           │
│            │  regardless of whether one is annotated, and pick the best one; for this     │
│            │  reason, "edit" may be slow                                                  │
│comp        │  a measure of CDR3 complexity, which is the total number of S, D and I       │
│            │  symbols in "edit" as defined above                                          │
├────────────┼──────────────────────────────────────────────────────────────────────────────┤
│cdr*_aa     │  the CDR*_AA sequence, or "unknown" if not computed                          │
│cdr*_len    │  number of amino acids in the CDR* sequence, or "unknown" if not computed    │
│cdr*_dna    │  the CDR*_DNA sequence, or "unknown" if not computed                         │
│fwr*_aa     │  the FWR*_AA sequence, or "unknown" if not computed                          │
│fwr*_len    │  number of amino acids in the FWR* sequence, or "unknown" if not computed    │
│fwr*_dna    │  the FWR*_DNA sequence, or "unknown" if not computed                         │
│            │  For all of these, * is 1 or 2 or 3 (or 4, for the fwr variables).           │
│            │  For CDR1 and CDR2, please see "enclone help amino" and the page on          │
│            │  <span style="color:#25bc24;">bit.ly/enclone</span> on V(D)J features.                                           │
├────────────┼──────────────────────────────────────────────────────────────────────────────┤
│vjlen       │  number of bases from the start of the V region to the end of the J region   │
│clen        │  length of observed constant region (usually truncated at primer start)      │
│ulen        │  length of observed 5'-UTR sequence;                                         │
│            │  note however that what report is just the start of the V segment            │
│            │  on the contig, and thus the length may include junk before the UTR          │
│cdiff       │  differences with universal reference constant region, shown in the          │
│            │  abbreviated form e.g. 22T (ref changed to T at base 22) or 22T+10           │
│            │  (same but contig has 10 additional bases beyond end of ref C region         │
│            │  At most five differences are shown, and if there are more, ... is appended. │
│udiff       │  like cdiff, but for the 5'-UTR                                              │
├────────────┼──────────────────────────────────────────────────────────────────────────────┤
│notes       │  optional note if there is an insertion or the end of J does not exactly abut│
│            │  the beginning of C; elided if empty; also single base overlaps between      │
│            │  J and C are not shown unless you use the special option <span style="font-weight:bold;">JC1</span>; we do this     │
│            │  because with some VDJ references, one nearly always has such an overlap     │
├────────────┼──────────────────────────────────────────────────────────────────────────────┤
│ndiff&lt;n>vj  │  number of base differences within V..J between this exact subclonotype and  │
│            │  exact subclonotype n                                                        │
│d_univ      │  distance from universal reference, more specifically,                       │
│            │  number of base differences within V..J between this exact                   │
│            │  clonotype and universal reference, exclusive of indels, the last 15         │
│            │  bases of the V and the first 15 bases of the J                              │
│d_donor     │  distance from donor reference,                                              │
│            │  as above but computed using donor reference                                 │
└────────────┴──────────────────────────────────────────────────────────────────────────────┘

<span style="background-color:#cbcccd;"> </span> <span style="color:#c23621;">●</span> These variables have some alternate versions, as shown in the table below:
<span style="background-color:#cbcccd;"> </span> 
<span style="background-color:#cbcccd;"> </span> ┌──────────┬───────────────────────────────┬──────────┬──────────────┬─────────────┬────────────┐
<span style="background-color:#cbcccd;"> </span> │variable  │  semantics                    │  visual  │  visual      │  parseable  │  parseable │
<span style="background-color:#cbcccd;"> </span> │          │                               │          │  (one cell)  │             │  (one cell)│
<span style="background-color:#cbcccd;"> </span> ├──────────┼───────────────────────────────┼──────────┼──────────────┼─────────────┼────────────┤
<span style="background-color:#cbcccd;"> </span> │x         │  median over cells            │  yes     │  this cell   │  yes        │  yes       │
<span style="background-color:#cbcccd;"> </span> │x_mean    │  mean over cells              │  yes     │  null        │  yes        │  yes       │
<span style="background-color:#cbcccd;"> </span> │x_μ       │  (same as above)              │  yes     │  null        │  yes        │  yes       │
<span style="background-color:#cbcccd;"> </span> │x_sum     │  sum over cells               │  yes     │  null        │  yes        │  yes       │
<span style="background-color:#cbcccd;"> </span> │x_Σ       │  (same as above)              │  yes     │  null        │  yes        │  yes       │
<span style="background-color:#cbcccd;"> </span> │x_min     │  min over cells               │  yes     │  null        │  yes        │  yes       │
<span style="background-color:#cbcccd;"> </span> │x_max     │  max over cells               │  yes     │  null        │  yes        │  yes       │
<span style="background-color:#cbcccd;"> </span> │x_%       │  % of total GEX (genes only)  │  yes     │  this cell   │  yes        │  yes       │
<span style="background-color:#cbcccd;"> </span> │x_cell    │  this cell                    │  no      │  no          │  no         │  this cell │
<span style="background-color:#cbcccd;"> </span> └──────────┴───────────────────────────────┴──────────┴──────────────┴─────────────┴────────────┘
<span style="background-color:#cbcccd;"> </span> Some explanation is required.  If you use enclone without certain options, you get the "visual"
<span style="background-color:#cbcccd;"> </span> column.
<span style="background-color:#cbcccd;"> </span> • Add the option <span style="font-weight:bold;">PER_CELL</span> (see "enclone help display") and then you get visual output with extra
<span style="background-color:#cbcccd;"> </span> lines for each cell within an exact subclonotype, and each of those extra lines is described by
<span style="background-color:#cbcccd;"> </span> the "visual (one cell)" column.
<span style="background-color:#cbcccd;"> </span> • If you generate parseable output (see "enclone help parseable"), then you get the "parseable"
<span style="background-color:#cbcccd;"> </span> column for that output, unless you specify <span style="font-weight:bold;">PCELL</span>, and then you get the last column.
<span style="background-color:#cbcccd;"> </span> • For the forms with μ and Σ, the Greek letters are only used in column headings for visual output
<span style="background-color:#cbcccd;"> </span> (to save space), and optionally, in names of fields on the command line.
<span style="background-color:#cbcccd;"> </span> <span style="color:#25bc24;">▶</span> If you try out these features, you'll see exactly what happens! <span style="color:#25bc24;">◀</span>

At least one variable must be listed.  The default is <span style="font-weight:bold;">u,const,notes</span>.  <span style="font-weight:bold;">CVARSP</span>: same as <span style="font-weight:bold;">CVARS</span> but
appends.

</span></pre>
</body>
</html>
