
[01minformation about how enclone works[0m

The goal of enclone is to find and display the clonotypes within single cell VDJ datasets: groups
of cells having the same fully rearranged common ancestor.

enclone provides the foundation for fully understanding each cell's antigen affinity and the
evolutionary relationship between cells within a sample.  This starts with, for each cell, [01mthe
full length sequence of all its VDJ receptor chains[0m.  Such data may be obtained using the 10x
Genomics immune profiling platform.

For this, there are fundamental challenges:

â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚1. It is extremely easy to get false positives: the incorrect appearance that two cells have a    â”‚
â”‚common ancestor.                                                                                  â”‚
â”‚                                                                                                  â”‚
â”‚2. Because of somatic hypermutation in B cells, it can be difficult to know that two B cells shareâ”‚
â”‚a common ancestor.                                                                                â”‚
â”‚                                                                                                  â”‚
â”‚3. There is always some background noise, e.g. from ambient mRNA.  When building large clonotypes,â”‚
â”‚this noise tends to pile up, yielding ectopic chains, i.e. chains within a clonotype that are     â”‚
â”‚artifacts and do not represent true biology.                                                      â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜

To address these challenges, the enclone algorithm has several steps, which we outline:

[01m[31m1[0m.  Input data.  enclone gets its information from the file all_contig_annotations.json that is
produced by Cell Ranger.  Only productive contigs are used.  Each has an annotated V and J
segment.  The V segment alignment may have a single indel whose length is divisible by three, and
in that case, the V reference sequence is edited either to delete or insert sequence.  In the
insertion case, the bases are taken from the contig.  These indels are noted in the enclone
output.

[01m[31m2[0m.  Exact subclonotypes.  enclone groups cells into exact subclonotypes, provided that they have
the same number of chains, identical V..J sequences, identical C segment assignments, and the same
distance between the J stop and the C start (which is usually zero).

[01m[31m3[0m.  Finding the germline sequences.  For samples from a given donor, enclone derives "donor
reference sequences" for the V chains present in the donor's genome.  This is powerful, even
though based on imperfect information.  V segments vary in their expression frequency and thus the
more cells which are present, the more complete the information will be.  It is also not possible
to accurately determine the last ~15 bases in a V chain from transcript data because these bases
are mutated during recombination.

[01m[31m4[0m.  What joins are tested.  Pairs of exact subclonotypes are considered for joining, as described
below.  This process only considers exact subclonotypes have two or three chains.  There is some
separate joining for the case of one chain.  Exact subclonotypes having four chains are not joined
at present.  These cases are clearly harder because these exact subclonotypes are highly enriched
for cell doublets, which we discard if we can identify as such.

[01m[31m5[0m.  Initial grouping.  For each pair of exact subclonotypes, and for each pair of chains in each
of the two exact subclonotypes, for which V..J has the same length for the corresponding chains,
and the CDR3 segments have the same length for the corresponding chains, enclone considers joining
the exact subclonotypes into the same clonotype.

[01m[31m6[0m.  Error bounding.  To proceed, as a minimum requirement, there must be at most 50 total
mismatches between the two exact subclonotypes, within the given two V..J segments.

[01m[31m7[0m.  Shared mutations.  enclone next finds shared mutations betweens exact subclonotypes, that is,
for two exact subclonotypes, common mutations from the reference sequence, using the donor
reference for the V segments and the universal reference for the J segments.  Shared mutations are
supposed to be somatic hypermutations, that would be evidence of common ancestry.  By using the
donor reference sequences, most shared germline mutations are excluded, and this is critical for
the algorithm's success.

[01m[31m8[0m.  Are there enough shared mutations?  We find the probability p that â€œthe shared mutations occur
by chanceâ€.  More specifically, given d shared mutations, and k total mutations (across the two
cells), we compute the probability p that an sample with replacement of k items from a set whose
size is the total number of bases in the V..J segments, yields at most k â€“ d distinct elements. 
The probability is an approximation, for the method please see
[32mhttps://docs.rs/stirling_numbers/0.1.0/stirling_numbers[0m.

[01m[31m9[0m.  Are there too many CDR3 mutations?  Next, let N be "the number of DNA sequences that differ
from the given CDR3 sequences by at most the number of observed differences".  More specifically,
if cd is the number of differences between the given CDR3 nucleotide sequences, and n is the total
length in nucleotides of the CDR3 sequences (for the two chains), we compute the total number N of
strings of length n that are obtainable by perturbing a given string of length n, which is
sum( choose(n,m), m = 0..=cd) ).

[01m[31m10[0m.  Key join criteria.  Two cells sharing sufficiently many shared differences and sufficiently
few CDR3 differences are deemed to be in the same clonotype.  That is, The lower p is, and the
lower N is, the more likely it is that the shared mutations represent bona fide shared ancestry. 
Accordingly, the smaller p*N is, the more likely it is that two cells lie in the same true
clonotype.  To join two cells into the same clonotype, we require that the bound p*n â‰¤ C is
satisfied, where C is the constant 1,000,000.  This constant was arrived at by empirically
balancing sensitivity and specificity across a large collection of datasets.  See discussion of
performance below.

[01m[31m11[0m.  Junk.  Spurious chains are filtered out based on frequency and connections. See "enclone help
special" for a description of the filters.

We are actively working to improve the algorithm.  To test the performance of the current version,
we combined data from 443 BCR libraries from 30 donors, which yielded [01m[31m9573[0m clonotypes having at
least two cells each, of which [01m[31m15 (0.16%)[0m contained data from multiple donors.  These are errors.

