
[01mglossary of terms used by enclone[0m

â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚V..J                â”‚  the full sequence of a V(D)J transcript, from the beginning of the V       â”‚
â”‚                    â”‚  segment to the end of the J segment; this sequence begins with a stop codonâ”‚
â”‚                    â”‚  and ends with a partial codon (its first base)                             â”‚
â”‚CDR3                â”‚  The terms CDR3 and junction are commonly mistaken and often used           â”‚
â”‚                    â”‚  interchangeably.  In enclone's nomenclature, "CDR3" actually refers to the â”‚
â”‚                    â”‚  junction (the CDR3 loop plus the canonical C and W/F at the N and C terminiâ”‚
â”‚                    â”‚  respectively).                                                             â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚clonotype           â”‚  all the cells descended from a single fully rearranged T or B cell         â”‚
â”‚                    â”‚  (approximated computationally)                                             â”‚
â”‚exact subclonotype  â”‚  all cells having identical transcripts [01m[31mâ—‹[0m                                   â”‚
â”‚                    â”‚  (every clonotype is a union of exact subclonotypes)                        â”‚
â”‚clone               â”‚  a cell in a clonotype, or in an exact subclonotype                         â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚onesie              â”‚  a clonotype or exact subclonotype having exactly one chain                 â”‚
â”‚twosie              â”‚  a clonotype or exact subclonotype having exactly two chains                â”‚
â”‚threesie            â”‚  a clonotype or exact subclonotype having exactly three chains;             â”‚
â”‚                    â”‚  these frequently represent true biological events, arising from expression â”‚
â”‚                    â”‚  of both alleles                                                            â”‚
â”‚foursie             â”‚  a clonotype or exact subclonotype having exactly four chains;              â”‚
â”‚                    â”‚  these very rarely represent true biological events                         â”‚
â”‚moresie             â”‚  a clonotype having more than four chains;                                  â”‚
â”‚                    â”‚  these sad clonotypes do not represent true biological events               â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚donor               â”‚  an individual from whom samples are obtained                               â”‚
â”‚sample              â”‚  a tube of cells from a donor, from a particular tissue at a                â”‚
â”‚                    â”‚  particular point in time, and possibly enriched for particular cells       â”‚
â”‚cell group          â”‚  an aliquot from a sample, presumed to be a random draw                     â”‚
â”‚dataset             â”‚  all sequencing data obtained from a particular library type                â”‚
â”‚                    â”‚  (e.g. TCR or BCR or GEX or FB), from one cell group, processed by running  â”‚
â”‚                    â”‚  through the Cell Ranger pipeline                                           â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜

[01m[31mâ—‹[0m The exact requirements for being in the same exact subclonotype are that cells:
â€¢ have the same number of productive contigs identified
â€¢ that these have identical bases within V..J
â€¢ that they are assigned the same constant region reference sequences
â€¢ and that the difference between the V stop and the C start is the same
  (noting that this difference is nearly always zero).
Note that we allow mutations within the 5'-UTR and constant regions.

[01mconventions[0m

â€¢ When we refer to "V segments", we always include the leader segment.
â€¢ Zero or one?  We number exact subclonotypes as 1, 2, ... and likewise with
chains within a clonotype, however DNA and amino-acid positions are numbered starting at zero.

