<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<!--  -->
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="application/xml+xhtml; charset=UTF-8"/>
<title>enclone help </title>
<link href='../enclone.css' rel='stylesheet' type='text/css'>
</head>
<body>
<pre style='font-family: "DejaVuSansMono"; line-height: 110%'><span style="font-size: 14px">
The mission of <span style="color:#3e7dd1;font-weight:bold;">e</span><span style="color:#f37721;font-weight:bold;">n</span><span style="color:#e7bd20;">c</span><span style="color:#72c0ff;font-weight:bold;">l</span><span style="color:#eb99c2;font-weight:bold;">o</span><span style="color:#31c39a;">n</span><span style="color:#f37721;font-weight:bold;">e</span> is to:

<span style="font-weight:bold;">  Find and display the clonotypes within single cell VDJ datasets:</span>
<span style="font-weight:bold;">  groups of cells having the same fully rearranged common ancestor.</span>

<span style="color:#5833ff;font-weight:bold;">enclone is part of the 10x Genomics immune profiling tools, including Cell Ranger and Loupe. 
enclone uses output from Cell Ranger version ≥ 3.1.</span>

The complete enclone documentation is at <span style="color:#25bc24;">bit.ly/enclone</span>.  This page catalogs the subset of those
pages that are directly accessible from the enclone command line.  These pages can be viewed in a
100 wide x 56 high window, except for those labeled "long" or "wide".

┌─────────────────────────┬─────────────────────────────────────────────────────────────────────┐
│<span style="font-weight:bold;">command</span>                  │  <span style="font-weight:bold;">what it provides</span>                                                   │
├─────────────────────────┼─────────────────────────────────────────────────────────────────────┤
│enclone help             │  help to test for correct setup                                     │
│enclone                  │  what you see here: guide to all the doc                            │
├─────────────────────────┼─────────────────────────────────────────────────────────────────────┤
│enclone help quick       │  quick guide to getting started                                     │
│enclone help how         │  how enclone works (long)                                           │
│enclone help command     │  info about enclone command line argument processing                │
├─────────────────────────┼─────────────────────────────────────────────────────────────────────┤
│enclone help glossary    │  glossary of terms used by enclone, and conventions                 │
├─────────────────────────┼─────────────────────────────────────────────────────────────────────┤
│enclone help example1    │  explanation of an example                                          │
│enclone help example2    │  example showing gene expression and feature barcodes (wide)        │
├─────────────────────────┼─────────────────────────────────────────────────────────────────────┤
│enclone help support     │  how we can help, enclone@10xgenomics.com                           │
├─────────────────────────┼─────────────────────────────────────────────────────────────────────┤
│enclone help input       │  how to provide input to enclone (long)                             │
│enclone help input_tech  │  how to provide input to enclone (technical notes)                  │
│enclone help parseable   │  parseable output (long)                                            │
├─────────────────────────┼─────────────────────────────────────────────────────────────────────┤
│enclone help filter      │  clonotype filtering options, scanning for feature enrichment (long)│
│enclone help special     │  special filtering options (long)                                   │
├─────────────────────────┼─────────────────────────────────────────────────────────────────────┤
│enclone help lvars       │  lead column options (long)                                         │
│enclone help cvars       │  per chain column options (long)                                    │
│enclone help amino       │  per chain column options for amino acids                           │
│enclone help display     │  other clonotype display options                                    │
├─────────────────────────┼─────────────────────────────────────────────────────────────────────┤
│enclone help indels      │  insertion and deletion handling                                    │
├─────────────────────────┼─────────────────────────────────────────────────────────────────────┤
│enclone help color       │  how enclone uses color, and related things                         │
│enclone help ideas       │  ideas for features that might be implemented                       │
│enclone help faq         │  frequently asked questions (long)                                  │
│enclone help developer   │  a few things for developers                                        │
├─────────────────────────┼─────────────────────────────────────────────────────────────────────┤
│<span style="color:#31c39a;">enclone help all</span>         │  <span style="color:#31c39a;">concatenation of all the help pages (long, wide)</span>                   │
│                         │  <span style="color:#31c39a;">███ USE THIS TO SEARCH ALL THE HELP PAGES! ███</span>                     │
└─────────────────────────┴─────────────────────────────────────────────────────────────────────┘
</span></pre>
</body>
</html>
