
[01m[31mFrequently Asked Questions[0m

We're sorry you're having difficulty!  Please see the answers below, check out the other help
guides, and if you're still stuck, write to us at enclone@10xgenomics.com.

[01m[38;5;12m1. Why is my enclone output garbled?[0m

We can think of two possibilities:

A. The escape characters that enclone emits for color and bolding are not getting
translated.  You have some options:
(a) Turn off escape character generation by adding PLAIN to your enclone commands.
This will work but you'll lose some information.
(b) If your terminal window is not translating escape characters, ask someone
with appropriate expertise to help you.  We have not observed this phenomenon,
but it should be fixable.
(c) If you're trying to view enclone output, with escape characters, using an editor,
that's probably not going to work well.

B. Perhaps enclone is emitting very wide lines.  Here are things you can do about this:
(a) Make your terminal window wider or reduce the font size.
(b) Identify the field that is very wide and use the column controls to remove that
field.  See the help for lvars and cvars.  For example,
[01mAMINO=cdr3[0m
may help, or even
[01mAMINO=[0m

[01m[38;5;12m2. Can I convert the enclone visual output into other forms?[0m

Yes, there are choices:
[01mA[0m. On a Mac, you can screenshot from a terminal window.
[01mB[0m. Add the argument [01mHTML[0m to the enclone command line.  Then the output will be presented as html.
[01mC[0m. You can then convert the html to pdf.  The best way on a Mac is to open Safari, which is the
best browser for this particular purpose, select the file where you've saved the html, and then
export as pdf.  Do not convert to pdf via printing, which produces a less readable file, and also
distorts colors.  (We do not know why the colors are distorted.)
[01mD[0m. If you want to put enclone output in a Google Doc, you can do it via approache [01mA[0m, although then
you won't be able to select text within the copied region.  Alternatively, if you open the html
file in a browser, you can then select text (including clonotype box text) and paste into a Google
Doc.  It will be pretty ugly, but will capture color and correctly render the box structure,
provided that you use an appropriate fixed-width font for that part of the Doc.  We found that
Courier New works, with line spacing set to 0.88.  You may have to reduce the font size.

[01m[38;5;12m3. Why is enclone slow for me?[0m

On a single VDJ dataset, it typically runs for us in a few seconds, on a Mac or Linux server. 
Runs where we combine several hundred datasets execute in a couple minutes (on a server).  Your
mileage could vary, and we are interested in cases where it is underperforming.  Let us know.  We
are aware of several things that could be done to speed up enclone.

[01m[38;5;12m4. What platforms does enclone run on?[0m

1. Linux/x86-64 (that's most servers)
2. Mac.

However, we have not and cannot test every possible configuration of these platforms.  Please let
us know if you encounter problems!

[01m[38;5;12m5. How can I print out all the donor reference sequences?[0m

Add the argument [01mDONOR_REF_FILE=filename[0m to your enclone command, and fasta for the donor
reference sequences will be dumped there.

[01m[38;5;12m6. How does enclone know what VDJ reference sequences I'm using?[0m

It does not!  It assumes that you have the [01mhuman[0m reference sequences that shipped with the latest
version of Cell Ranger.  If instead your sequences are mouse, then you can specify that by adding
the argument [01mMOUSE[0m to your command line.  If you are simply using another reference sequence,
please specify that by adding [01mREF=f[0m to your command line, where [01mf[0m is the name of your VDJ
reference fasta file.

[01m[38;5;12m7. Can I provide data from more than one donor?[0m

Yes.  Type [01menclone help input[0m for details.  The default behavior of enclone is to prevent cells
from different donors from being placed in the same clonotype.  The [01mMIX_DONORS[0m option may be used
to turn off this behavior.  If you employ this option, then clonotypes containing cells from more
than one donor will be flagged as errors, unless you use the [01mNWARN[0m option to turn off those
warnings.  The primary reason for allowing entry of data from multiple donors is to allow
estimation of enclone's error rate.

[01m[38;5;12m8. What are some command line argument values quoted?[0m

Command line argument values that contain any of these characters ;|* need to be quoted like so
[01mTCR="a;b"[0m
to prevent the shell from interpreting them for a purpose completely unrelated to enclone.  This
is a trap, because forgetting to add the quotes can result in nonsensical and confusing behavior!

[01m[38;5;12m9. If enclone fails, does it return nonzero exit status?[0m

Yes, unless output of enclone is going to a terminal.  In that case, you'll always get zero.

[01m[38;5;12m10. Could a cell be missing from an enclone clonotype?[0m

Yes, some cells are deliberately deleted.  The cell might have been deleted by one of the filters
described in [01menclone help special[0m, and which you can turn off.  We also delete cells for which
more than four chains were found.

[01m[38;5;12m11. Can enclone print summary stats?[0m

Yes, if you add the option [01mSUMMARY[0m, then some summary stats will be printed.  If you only want to
see the summary stats, then also add the option [01mNOPRINT[0m.

[01m[38;5;12m12. What is the notes column?[0m

The notes column appears if one of two relatively rare events occurs:

1. An insertion is detected in a chain sequence, relative to the reference.

2. The end of the J segment on a chain sequence does not exactly coincide with
   the beginning of the C segment.
The latter could correspond to one of several phenomena:
a. A transcript has an insertion between its J and C segments.
   This can happen.  See e.g. Behlke MA, Loh DY.
   Alternative splicing of murine T-cell receptor beta-chain transcripts.
   Nature 322(1986), 379-382.
b. There is an error in a reference sequence segment.
   We have tried to eliminate all such errors from the built-in references for
   human and mouse.
c. A cell produced a nonstandard transcript and also standard ones, and the
   Cell Ranger pipeline just happened to pick a nonstandard one.
d. There was a technical artifact and the sequence does not actually represent
   an mRNA molecule.

Overlaps of length exactly one between J and C segments are not shown unless you specify the
option [01mJC1[0m.  The reason for this is that certain reference sequences (notably those from IMGT and
those supplied with Cell Ranger 3.1) often have an extra base at the beginning of their C
segments, resulting in annoying overlap notes for a large fraction of clonotypes.

[01m[38;5;12m13. Can I cap the number of threads used by enclone?[0m

You can use the command-line argument [01mMAX_CORES=n[0m to cap the number of cores used in parallel
loops.  The number of threads used is typically one higher.

[01m[38;5;12m14. Does enclone work under Windows?[0m

No.  There are nontrivial technical problems with getting this to work.  If you're sufficiently
curious, see the notes in the source code file misc1.rs.  Please let us know if you're interested
in support for Windows.

[01m[38;5;12m15. Can I use enclone if I have only gene expression data?[0m

Possibly.  In some cases this works very well, but in other cases it does not.  Success depends on
dataset characteristics that have not been carefully investigated.  To attempt this, you need to
invoke Cell Ranger on the GEX dataset as if it was a VDJ dataset, and you need to specify to Cell
Ranger that the run is to be treated as BCR or TCR.  Two separate invocations can be used to get
both.  Note also that Cell Ranger has been only minimally tested for this configuration and that
this is not an officially supported Cell Ranger configuration.

[01m[38;5;12m16. How can I cite enclone?[0m

This version of enclone has been provided under a non-disclosure agreement,
however once enclone has officially launched, you will be able to cite this version as:
[32m10x Genomics, https://github.com/10XGenomics/enclone,
version dj/23 : a65fc0a+ : 2020-04-19 : release : linux : x86_64.
[0mAt some point subsequent to that, there will be a white paper to which you can refer, in addition
to a DOI minted at Zenodo.  In the spirit of reproducibility, you should provide the arguments
that you used when you ran enclone and indicate the version of Cell Ranger that you used to
generate the input data.

[01m[38;5;12m17. Can I print the enclone version?[0m

Yes, type "enclone version".

