<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<!--  -->
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="application/xml+xhtml; charset=UTF-8"/>
<title>enclone help special</title>
<style type="text/css">
@font-face {
    font-family: "DejaVuSansMono";
    src:
    local("Menlo Regular"),
    url('https://cdn.jsdelivr.net/npm/@deathbeds/jupyterlab-font-dejavu-sans-mono@1.0.0/style/fonts/DejaVuSansMono.woff2')
    format("woff2"),
    url('https://cdn.jsdelivr.net/npm/dejavu-fonts-ttf@2.37.3/ttf/DejaVuSansMono.ttf')
    format("truetype"),
    url('../fonts/DejaVuLGCSansMono.ttf') format("truetype"),
    local("Courier New");
}
@font-face {
    font-family: "DejaVuSansMono";
    font-weight: bold;
    src:
    local("Menlo Bold"),
    url('https://cdn.jsdelivr.net/npm/@deathbeds/jupyterlab-font-dejavu-sans-mono@1.0.0/style/fonts/DejaVuSansMono-Bold.woff2')
    format("woff2"),
    url('https://cdn.jsdelivr.net/npm/dejavu-fonts-ttf@2.37.3/ttf/DejaVuSansMono-Bold.ttf')
    format("truetype"),
    url('../fonts/DejaVuLGCSansMono-Bold.ttf') format("truetype"),
    local("Courier New Bold");
}
</style>
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-58278925-3"></script>
<script>
window.dataLayer = window.dataLayer || [];
function gtag(){{dataLayer.push(arguments);}}
gtag('js', new Date());
gtag('config', 'UA-58278925-3');
</script>
</head>
<body>

<br>
<img src="../../img/enclone_banner.png" alt="enclone banner" title="enclone banner" width=100% />
<pre style='font-family: "DejaVuSansMono"; line-height: 110%'><span style="font-size: 14px">
<span style="font-weight:bold;">special filtering options</span>

This page documents some options, most of which allow noise filters to be turned off, and which
normally should not be invoked.  The last two options can be used to simplify the view of a
clonotype.

┌────────────────────┬─────────────────────────────────────────────────────────────────────────────┐
│NALL                │  Turn off all the noise filters shown below.  This may yield quite a mess.  │
├────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
│NCELL               │  Use contigs found by Cell Ranger even if they were not in a called cell,   │
│                    │  or not called high confidence.                                             │
│NALL_CELL           │  turn off all the noise filters except for the cell filter                  │
├────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
│NGEX                │  If gene expression and/or feature barcode data are provided, if a barcode  │
│                    │  is called a cell by the VDJ part of the Cell Ranger pipeline, but not      │
│                    │  called a cell by the gene expression and/or feature barcode part, then the │
│                    │  default behavior of enclone is to remove such cells from clonotypes.  This │
│                    │  option disables that behavior.                                             │
├────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
│NCROSS              │  If you specify that two or more libraries arose from the same origin (i.e. │
│                    │  cells from the same tube or tissue), then by default enclone will          │
│                    │  "cross filter" so as to remove expanded exact subclonotypes that are       │
│                    │  present in one library but not another, in a fashion that would be highly  │
│                    │  improbable, assuming random draws of cells from the tube.  These are       │
│                    │  believed to arise when a plasma or plasmablast cell breaks up during during│
│                    │  or after pipetting from the tube, and the resulting fragments seed GEMs,   │
│                    │  yielding expanded 'fake' clonotypes that are residues of real single plasma│
│                    │  cells.  The NCROSS options turns off this filter, which could be useful so │
│                    │  long as you interpret the restored clonotypes as representing what are     │
│                    │  probably single cells.  There may also be other situations where the filter│
│                    │  should be turned off, and in particular the filter can do weird things if  │
│                    │  inputs are somehow mis-specified to enclone.  Note that for purposes of    │
│                    │  this option, enclone defines an origin by the pair                         │
│                    │  (origin name, donor name).                                                 │
├────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
│NUMI                │  Filter out B cells based on low BCR UMI counts.  The heuristics for this   │
│                    │  are described on the enclone site at <span style="color:#25bc24;">bit.ly/enclone</span>.                       │
│NUMI_RATIO          │  Filter out B cells based on low BCR UMI counts relative to another cell    │
│                    │  in a given clonotype.  The heuristics for this                             │
│                    │  are described on the enclone site at <span style="color:#25bc24;">bit.ly/enclone</span>.                       │
├────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
│NGRAPH_FILTER       │  By default, enclone filters to remove exact subclonotypes that by virtue of│
│                    │  their relationship to other exact subclonotypes, appear to arise from      │
│                    │  background mRNA or a phenotypically similar phenomenon.  The               │
│                    │  NGRAPH_FILTER option turns off this filtering.                             │
├────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
│NQUAL               │  By default, enclone filters out exact subclonotypes having a base in V..J  │
│                    │  that looks like it might be wrong.  More specifically, enclone finds bases │
│                    │  which are not Q60 for a barcode, not Q40 for two barcodes, are not         │
│                    │  supported by other exact subclonotypes, are variant within the clonotype,  │
│                    │  and which disagree with the donor reference.  NQUAL turns this off.        │
├────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
│NWEAK_CHAINS        │  By default, enclone filters chains from clonotypes that are                │
│                    │  weak and appear to be artifacts, perhaps arising from a stray mRNA molecule│
│                    │  that floated into a GEM.  The NWEAK_CHAINS option turns off this filter.   │
├────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
│NWEAK_ONESIES       │  By default, enclone filters out onesie clonotypes having a single exact    │
│                    │  subclonotype, and that are light chain or TRA, and whose number of cells is│
│                    │  greater than one but less than 0.1% of the total number of cells.          │
│                    │  This filter reduces the likelihood of creating clonotypes containing cells │
│                    │  that arose from different recombination events.                            │
│                    │  NWEAK_ONESIES turns this filter off.                                       │
├────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
│NFOURSIE_KILL       │  By default, if enclone finds a foursie exact subclonotype that             │
│                    │  contains a twosie exact subclonotype having at least ten cells, it kills   │
│                    │  the foursie exact subclonotype, no matter how many cells it has.  The      │
│                    │  foursies that are killed are believed to be rare oddball artifacts arising │
│                    │  from repeated cell doublets or GEMs that contain two cells and multiple gel│
│                    │  beads.  The argument NFOURSIE_KILL turns off this filtering.               │
├────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
│NWHITEF             │  By default, enclone filters out rare artifacts arising from contamination  │
│                    │  of oligos on gel beads.  The NWHITEF option turns off this filter.         │
├────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
│NBC_DUP             │  By default, enclone filters out duplicated barcodes within an exact        │
│                    │  subclonotype.  The NBC_DUP option turns off this filter.                   │
├────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
│MIX_DONORS          │  By default, enclone will prevent cells from different donors from being    │
│                    │  placed in the same clonotype.  The MIX_DONORS option turns off this        │
│                    │  behavior, thus allowing cells from different donors to be placed in the    │
│                    │  same clonotype.  The main use of this option is for specificity testing, in│
│                    │  which data from different donors are deliberately combined in an attempt   │
│                    │  to find errors.  Use of the bc field for META input specification          │
│                    │  automatically turns on this option.                                        │
├────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
│NIMPROPER           │  enclone filters out exact subclonotypes having only one chain type (heavy  │
│                    │  or light or TRA or TRB).  (Exception: it does not remove clonotypes having │
│                    │  exactly one chain, but whose number of cells is at least (total            │
│                    │  cells)/10000.) For example, the filter removes all exact subclonotypes     │
│                    │  having two TRA chains and no other chains.  The NIMPROPER option turns off │
│                    │  this filter.                                                               │
├────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
│MIN_CHAINS_EXACT=n  │  Delete any exact subclonotype having less than n chains.  You can use this │
│                    │  to "purify" a clonotype so as to display only exact subclonotypes having   │
│                    │  all their chains.                                                          │
│CHAINS_EXACT=n      │  Delete any exact subclonotype not having exactly n chains.                 │
│MIN_CELLS_EXACT=n   │  Delete any exact subclonotype having less than n cells.  You might want    │
│                    │  to use this if you have a very large and complex expanded clonotype,       │
│COMPLETE            │  delete any exact subclonotype that has less chains than the clonotype      │
│                    │  for which you would like to see a simplified view.                         │
├────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
│FCELL=var=value     │  Supposing that var has been specified as a field using the BC option       │
│                    │  (or equivalently, using bc, via META), see "enclone help input", this      │
│                    │  option filters out all barcodes that do not satisfy the given constraint.  │
│                    │  Note that for purposes of testing the constraint, if the value for a       │
│                    │  particular barcode has not been specified via BC or bc, then its value is  │
│                    │  taken to be null.  Also multiple instances of FCELL may be used to impose  │
│                    │  multiple filters.                                                          │
└────────────────────┴─────────────────────────────────────────────────────────────────────────────┘

</span></pre>
</body>
</html>
