<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" 
"https://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<!--  -->
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="application/xml+xhtml; charset=UTF-8"/>
<title></title>
<link rel="stylesheet" type="text/css" href="../enclone_css_v2.css">
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-58278925-3"></script>
<script>
window.dataLayer = window.dataLayer || [];
function gtag(){{dataLayer.push(arguments);}}
gtag('js', new Date());
gtag('config', 'UA-58278925-3');
</script>

        
        </head>

        <! â€“â€“

        ðŸ’© ðŸ’© ðŸ’© ðŸ”´ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”´ ðŸ’© ðŸ’© ðŸ’©

        PUT DOWN YOUR HAMMER.
        THIS IS AN AUTO-GENERATED FILE.  PLEASE DO NOT EDIT IT.
        THANK YOU FOR YOUR COOPERATION,

        SINCERELY,
        THE BENEVOLENT OVERLORDS

        ðŸ’© ðŸ’© ðŸ’© ðŸ”´ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”´ ðŸ’© ðŸ’© ðŸ’©

        â€“â€“>
<body>

<br>
<a href="../../index.html#help">
<img src="../../img/enclone_banner.png" alt="enclone banner" title="enclone banner" width=100% />
</a>

<h1>History</h1>

<p>This page provides a selective history of what was changed in <span style="color:rgb(120,123,175);font-weight:900">enclone</span> and when.</p>

<p>
We show changes that affect users like new features, changes to results, and the like.  This
log starts with with initial public availability.  The complete history may be seen by cloning the 
<span style="color:rgb(120,123,175);font-weight:900">enclone</span> repo and typing <code>git log</code>.
</p>

<p><span style="color:red">Breaking changes are shown in red.</span></p>

<p style="border: 2px; border-style: solid; border-color: black; background-color: #EBF4FC;
padding: 8px; width: 900px; font-size: 110%">
Please be aware that our workflow when we make changes is to automatically update the GitHub
site, including all the website pages and this page too.  This happens <i>in advance</i> of 
actually making a release (which might follow in a couple days).  This means that the website
may describe features that are not yet available in a release (although they will be in the
source code that's available).  We apologize for this asynchrony!  Note however that the 
command-line help that comes with your copy of <span style="color:rgb(120,123,175);font-weight:900">enclone</span> will always match its behavior.</p>

<hr>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p><b>9/21/21</b>: 
<ul>
<li>Simplify the install script and expand the installation debugging page.  This may resolve some 
problems people have had and also provides help in the event that installation fails.</li>
<li>Add new lead variable <code>nchains_present</code>.  Please see
<a href="../../pages/auto/help.lvars.html"><code>enclone help lvars</code></a>.</li>
</ul>
</p>

<p><b>9/20/21</b>: 
Add a new default filter called <i>signature filtering</i> that under certain rare circumstances
will decompose complex clonotypes that are incorrectly glued together.  This filter can 
dramatically affect certain datasets, but has almost no effect on typical data (less than one per 
million clonotypes tested).
Please see <a href="../../pages/auto/default_filters.html">enclone default filters</a>.</p>

<p><b>9/17/21</b>: 
<ul>
<li>Add <code>KEEP_CLONO_IF_CELL_MIN</code>.  See
<a href="../../pages/auto/help.filter.html"><code>enclone help filter</code></a>.</li>
<li>Tweak handling of <code>NOPRINT PARSEABLE=stdouth</code> to make columns line up.</li>
</ul>
</p>

<p><b>9/15/21</b>: Add control over PNG resolution for cell coloring by variable.</p>

<p>
<b>9/14/21</b>: Fix some issues with cell coloring, affecting cell coloring by variable and
<code>SPLIT_PLOT_BY_ORIGIN</code>.
</p>

<p>
<b>9/11/21</b>: Add the ability to color cells in a honeycomb plot using the values of a variable.
Please see <a href="../../pages/auto/plot.html">enclone plots</a>.
</p>

<p>
<b>8/27/21</b>: Add new argument <code>DVARS=...</code> that can display dataset-level variables,
for now very limited, as described at
see <a href="../../pages/auto/help.display.html"><code>enclone help display</code></a>.
In addition, values for <code>GVARS</code> now appear inside the summary.
</p>

<p>
<b>8/24/21</b>: 
<ul>
<li><code>SPLIT_PLOT_BY_ORIGIN</code> now allows creation of side-by-side honeycomb
plots, representing e.g. different cell types or preparations.  See
<a href="../../pages/auto/plot.html">enclone plots</a>.</li>
<li><span style="color:rgb(120,123,175);font-weight:900">enclone</span> installation now includes a test for whether the correct <span style="color:rgb(120,123,175);font-weight:900">enclone</span> executable is
accessible from the command line, and prints debugging information if not.</li>
</ul>
</p>

<p>
<b>8/17/21</b>: <code>KEEP_CELL_IF</code> now allows feature variables to appear in constraints, 
see <a href="../../pages/auto/help.special.html"><code>enclone help special</code></a>.
</p>

<p>
<b>7/22/21</b>:
<ul>
<li>Add option for xy plots to make them symmetric, which is appropriate e.g. for plotting
data from replicates.  See <a href="../../pages/auto/plot.html">enclone plots</a>.</li>
<li>Add option <code>INFO_RESOLVE</code> that modifies <code>INFO</code> to handle cases where
identical immune receptor sequences have been provided.  See
<a href="../../pages/auto/help.input.html"><code>enclone help input</code></a>.
</li>
</ul>

<p>
<b>7/14/21</b>: Add new arguments <code>HONEY_OUT</code> and <code>HONEY_IN</code> to allow
preservation of cell positions between two honeycomb plots of the same data.  Please see
<a href="../../pages/auto/plot.html">enclone plots</a>.
</p>

<p>
<b>6/30/21</b>: Add cvar <code>cigar</code> that is the CIGAR string of the V..J contig sequence
versus the universal reference.  See
<a href="../../pages/auto/help.cvars.html"><code>enclone help cvars</code></a>.
</p>

<p>
<b>6/24/21</b>: Now <code>enclone UPDATE</code> updates to the latest version of <span style="color:rgb(120,123,175);font-weight:900">enclone</span>.
</p>

<p>
<b>6/17/21</b>: Add <code>SIM_MAT_PLOT</code>, to display the pairwise cosine similarity between
a list of variables, see <a href="../../pages/auto/plot.html">enclone plots</a>.
</p>

<p>
<b>6/4/21</b>: Tweak the order of default filters so as to increase the likelihood that clonotypes
that are held together tenuously (or not at all) will be pulled apart.  This has a tiny effect
on our datasets (changing about <code>1</code> in <code>10^5</code> clonotypes), but could affect
some user datasets disproportionally.
</p>

<p>
<b>6/3/21</b>: Fix bug in the computation of <code>datasets_cell</code>.
</p>

<p>
<b>6/2/21</b>: 
<ul>
<li>Allow <code>PCHAINS=max</code>.  See
    <a href="../../pages/auto/help.parseable.html"><code>enclone help parseable</code></a>.</li>
<li>Add <code>MAX_HEAVIES=1</code> to ignore any cell having more than one IGH or TRB chain.</li>
<li>Fix bug that caused VDJ gene name and id fields to be displayed in parseable output even
    when a chain was absent for a given exact subclonotype.</li>
</ul>
</p>

<p>
<b>5/6/21</b>: Clonotype grouping has been completely reworked.  Please see
    <a href="../../pages/auto/help.display.html"><code>enclone help display</code></a>.</li>
The pre-existing options are still present but now undocumented, and their behavior might have
changed slightly.  We have not optimized the new options for run time, so let us know if there's
a problem and we'll see what we can do.
</p>

<p>
<b>5/4/21</b>: Several changes that involve D gene assignment, and which allow for the VDDJ case:
<ul>
<li>Add cvars <code>d1_name</code>, <code>d2_name</code>, <code>d1_score</code>, 
    <code>d2_score</code>, <code>d_delta</code> or equivalently <code>d_Î”</code>, see
    <a href="../../pages/auto/help.cvars.html"><code>enclone help cvars</code></a> and
    <a href="../../pages/auto/d_genes.html"><code>D genes and junction regions</code></a>.</li>
<li>Add options <code>ALIGN&ltn></code> and <code>JALIGN&ltn></code> to exhibit visual 
    alignments of V..J sequences or just the junction regions to the concatenated VDJ reference, see
    <a href="../../pages/auto/help.display.html"><code>enclone help display</code></a> and
    <a href="../../pages/auto/d_genes.html"><code>D genes and junction regions</code></a>.
    There are also <code>ALIGN_2ND&ltn></code> and <code>JALIGN_2ND&ltn></code> to use instead the 
    second best D gene alignments.</li>
<li>Add global variables <code>d_inconsistent_%</code> and <code>d_inconsistent_n</code> that
    provide statistics on overall consistency of D gene assignment, see
    <a href="../../pages/auto/help.display.html"><code>enclone help display</code></a> and
    <a href="../../pages/auto/d_genes.html"><code>D genes and junction regions</code></a>.</li>
<li>Add filters <code>D_INCONSISTENT</code> to only show clonotypes having an inconsistent D gene 
    assignment, <code>D_NONE</code> to show clonotypes have a null D gene assignment, and
    <code>D_SECOND</code> to show only VDDJ clonotypes.</li>
<li>Add grouping options <code>GROUP_VJ_REFNAME_HEAVY</code> and <code>GROUP_VDJ_REFNAME_HEAVY</code>, see
    <a href="../../pages/auto/help.display.html"><code>enclone help display</code></a>.</li>
</ul>
Additionally, the main page on this site now provides links to previous releases of <span style="color:rgb(120,123,175);font-weight:900">enclone</span>
that match particular releases of Cell Ranger.  And there is a new filter <code>MAX_EXACTS</code>.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>4/8/21</b>: 
<ul>
<li>Fix bug that caused <code>CONP</code> to produce the wrong answer in some cases.</li>
<li>Fix bug that caused FWR4 values to be wrong in some cases.</li>
</ul>
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>4/5/21</b>: Add page
<a href="../../pages/auto/default_filters.html"><code><span style="color:rgb(120,123,175);font-weight:900">enclone</span> default filters</code></a>
that describes the order of the filters that <span style="color:rgb(120,123,175);font-weight:900">enclone</span> uses, and provides technical details
about some of the filters, which was previously on the
<a href="../../pages/auto/heuristics.html"><code>heuristics page</code></a>.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>4/2/21</b>: 
<ul>
<li>Almost any parseable variable can now be used in a linear condition.  Of course it doesn't
    makes sense to use such a variable unless its values are numeric.</n>
<li>A number of other restrictions on variables were removed, including by promotion of parseable
    variables to lvars or cvars.</li>
<li>Clicking on the <span style="color:rgb(120,123,175);font-weight:900">enclone</span> banner at the top of any page on the site now takes you back to
    the help section on the main page.</li>
<li>There is a new <a href="../../pages/auto/variables.html"><code>page</code></a>
    that inventories all the <span style="color:rgb(120,123,175);font-weight:900">enclone</span> variables.</li>
<li>We turned on overflow checking in the <span style="color:rgb(120,123,175);font-weight:900">enclone</span> executables.  It is possible that you will
    now see error messages (tracebacks) where you did not previously see them.  Please report 
    these events!</li>
</ul>
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>3/23/21</b>: 
<ul>
<li>Add new option top plot two variables, see bottom of 
<a href="../../pages/auto/plot.html">enclone plots</a>.</li>
<li>Add <code>SOURCE=filename</code> to import some arguments from another file.</li>
<li>Add option <code>SCAN_EXACT</code> to scan exact subclonotypes rather than clonotypes.
See <a href="../../pages/auto/help.filter.html"><code>enclone help filter</code></a>.</li>
</ul>
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>3/17/21</b>: 
<ul>
<li>Gene scanning now has an option <code>SCAN_EXACT</code> to scan using exact subclonotypes
rather than clonotypes.  See
<a href="../../pages/auto/help.filter.html"><code>enclone help filter</code></a>.</li>
<li>Add <code>SOURCE=filename</code> to read extra arguments from the named file, see
<a href="../../pages/auto/help.command.html"><code>enclone help command</code></a>.</li>
</ul>
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>3/16/21</b>: 
<ul>
<li><span style="color:rgb(120,123,175);font-weight:900">enclone</span> can now take values for fields associated to particular immune receptor
sequences via <code>INFO=path.csv</code>; see
<a href="../../pages/auto/help.input.html"><code>enclone help input</code></a>.</li>
<li>We now allow <code>TREE=v1,...vn</code> where the <code>vi</code> are parseable variables,
resulting in labeling by these variables. See
<a href="../../pages/auto/tree.html">here</a>.  For backward compatibility, <code>const</code>
is translated to <code>const1</code>.  The new display is slightly different than the old
display, showing e.g. <code>const1=IGHG1</code> rather than <code>IGHG1</code>.
</li>
<li>There is now a detailed
<a href="../../pages/auto/dev.html"><code>developers guide</code></a>.</li>
</ul>
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>3/4/21</b>: 
<ul>
<li>
Add a plotting option <code>QUAD_HIVE</code> that forces honeycomb plots into the first
quadrant.  See <a href="../../pages/auto/plot.html">honeycomb plots</a>.
</li>
<li>
Tweak honeycomb plotting to cause smaller clonotypes to be generally "farther out" than
smaller clonotypes.  See <a href="../../pages/auto/plot.html">honeycomb plots</a>.
</li>
<li>
Add new options <code>CONX</code> and <code>CONP</code> that add amino acid consensus
rows to clonotype tables.  Please see
<a href="../../pages/auto/help.display.html"><code>enclone help display</code></a>, which refers to
<a href="../../pages/auto/help.cvars.html"><code>enclone help cvars</code></a>.
</li>
<li>For developers, add an option <code>PROFILE</code>.</li>
</ul>
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>2/14/21</b>: A VDJ reference file can now be tested for defects using a command of the form
<code>enclone REF=&ltfasta file name></code>.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>2/13/21</b>: 
<ul>
<li>
We redesigned the algorithm that determines the columns in clonotype tables,
see <a href="../../pages/auto/heuristics.html">notes on heuristics</a>.  This addresses some
very rare cases where chains were incorrectly placed in the same column.  This also results in
better disintegration of "merged" clonotypes in some cases.  There is an additional
change that came about.  In rare cases, a cell will have two contigs with identical CDR3 sequences.
This could be the "other" allele or more likely an artifact.  Previously one of these contigs
was suppressed in the clonotype table.  Now they both appear.
</li>
<li>
For weak chain filtering, we now allow deletion of chains having up to <code>20</code> cells, 
rather than <code>5</code>.  See <a href="../../pages/auto/heuristics.html">notes on heuristics</a>.
This fixes some pathological examples involving clonotypes having many chains.
</li>
<li>
Added option <code>FOLD_HEADERS</code> to reduce horizontal space usage.  See also the discussion
of line wrapping in
<a href="../../pages/auto/help.faq.html"><code>frequently asked questions</code></a>.
</li>
</ul>
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>2/6/21</b>: Fix the computations of median of a vector.  For a vector <code>x</code> of length 
<code>n</code>, these were computed as <code>x[n/2]</code>, which is incorrect in the case where
<code>x</code> has an even number of elements: it should be the mean of the innermost two
elements.  However, in cases where the vector elements are integers, we now round this mean up to
the nearest integer, yielding a "rounded median", as the advantage of presenting the exact
median is outweighed by the loss of readability resulting from adding <code>.5</code> or 
<code>.0</code> to the display.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>2/4/21</b>: Add option <code>SUPPRESS_ISOTYPE_LEGEND</code>.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>2/3/21</b>: Add chain variable <code>cdr3_aa_conx</code> that is the consensus of the
CDR3 amino acid sequences across the clonotype, showing <code>X</code> for each variant residue,
and <code>cdr3_aa_conp</code>, which instead of <code>X</code> shows an amino acid property
symbol.  See <a href="../../pages/auto/help.cvars.html"><code>enclone help cvars</code></a>.
We thank Albert Vilella for suggesting these features.
</p>

<hr>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>1/19/21</b>: Add lead variables <code>count_cdr1_&ltregex></code> and similar, see
<a href="../../pages/auto/help.lvars.html"><code>enclone help lvars</code></a>.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>1/14/21</b>: Add chain variables corresponding to extended or trimmed versions of CDR
sequences, and North versions.  Please see
<a href="../../pages/auto/help.cvars.html"><code>enclone help cvars</code></a>.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>1/13/21</b>: 
<ul>
<li>Move onesie merger to a later stage in the process, after certain "junk" deletion steps,
to increase the likelihood of merges happening.  This significantly increases the number of
merges, so that after this change, many more clonotypes contain onesies.</li>
<li>The <code>BC</code> argument now may be either a <code>CSV</code> or <code>TSV</code> file.</li>
<li>Automatically find <code>multi/vdj_b</code> and <code>multi/vdj_t</code> subdirectories of 
outs.</li>
<li>Allow <code>BCR_GEX=p</code> as a shorthand for <code>BCR=p</code> and <code>GEX=p</code>, which
simplifies syntax when a <code>cellranger</code> multi run is used as input; ditto for
<code>TCR_GEX</code>.</li>
</ul>
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>1/9/21</b>: Following a suggestion of Ganesh Phad, single-cell clonotypes in honeycomb
plots are now localized by color.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>1/7/21</b>: 
<ul>
<li>Greatly speed up honeycomb plots for large numbers of clonotypes.</li>
<li>Change the installation command to use <code>bash</code> rather than <code>sh</code> to avoid 
failures resulting from <code>sh</code> being instantiated as a different shell.</li>
</ul>
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>1/4/21</b>: 
<ul>
<li>
Fix the output of the parseable variable <code>cdr3_start&lti></code> when there was 
an insertion earlier in the transcript.
</li>
<li>
Add an option <code>PLOT_BY_ISOTYPE_COLOR</code> to allow color specification for the
<code>PLOT_BY_ISOTYPE</code> argument.
</li>
</ul>
</p>

<hr>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>12/14/20</b>: Add option <code>METAX</code> that allows meta information to be provided on
the command line, rather than through a separate file.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>12/9/20</b>: 
<ul>
<li>Delete the <code>_ag</code> antigen feature label (which was documented at
<a href="../../pages/auto/help.lvars.html"><code>enclone help lvars</code></a>) because this
was in fact never supported by <code>cellranger</code>.</li>
<li>Change the algorithm for deciding to use a donor reference allele, so that it checks all 
donor reference alleles for all V genes having the same name as the one originally assigned to
a contig.  In some cases this changes results.  Note that this change comes after Cell Ranger
5.0 and with high probability will be integrated into the next major Cell Ranger release.</li>
<li> Change the default value of <code>MAX_DIFFS</code> from <code>50</code> to <code>55</code>.  
Change <code>MAX_DEGRADATION</code> from <code>3</code> to <code>2</code>.  Change the default
value for <code>MAX_SCORE</code> from <code>1,000,000</code> to <code>500,000</code>.  Change
the default value for <code>MAX_CDR3_DIFFS</code> from <code>10</code> to <code>15</code>.
Ditto comments about Cell Ranger.</li>
<li>Allow up to 5 (instead of 4) CDR3 amino acid differences when joining chains.</li>
<li>Find input data correctly if the <code>cellranger multi</code> pipeline is used.</li>
<li>Add a new filter that removes some doublets, and which can be turned off by supplying the
argument <code>NDOUBLET</code> (see
<a href="../../pages/auto/help.special.html"><code>enclone help special</code></a> and
<a href="../../pages/auto/heuristics.html">notes on heuristics</a>).</li>
</ul>
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>12/5/20</b>: 
<ul>
<li>
The argument <code>F</code> has been renamed to
<code>KEEP_CLONO_IF_CELL_MEAN</code>, although for backward compatibility, <code>F</code>
still works.  We have added the similar <code>KEEP_CLONO_IF_CELL_MAX</code>.  See
<a href="../../pages/auto/help.filter.html"><code>enclone help filter</code></a> for both.
</li>
<li>
Rename <code>FCELL</code> to <code>KEEP_CELL_IF</code>, but retain <code>FCELL</code> for
backward compatibility.
</li>
</ul>
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>12/2/20</b>: CDR3 sequences are now recomputed by <span style="color:rgb(120,123,175);font-weight:900">enclone</span>, rather than taken from the value
in <code>all_contig_annotations.json</code>.  This will result in changed values at a very low
frequency, and only if the CDR3 calculation algorithm was changed between the time of the relevant
<code>cellranger</code> release and the time of the <span style="color:rgb(120,123,175);font-weight:900">enclone</span> release.
</p>

<hr>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>11/24/20</b>: 
<ul>
<li>
Allow arbitrarily long CDR3 sequences.  This change is not part of Cell Ranger 5.0, but may
be part of a subsequent release.  In advance of that, in order to exploit this feature, you
will need to add the option <code>REPROD</code> to the <span style="color:rgb(120,123,175);font-weight:900">enclone</span> command line, to cause productive
pairs to be recomputed.  Allowing very long CDR3 sequences likely introduces some artifacts.  For
example, in human BCR data, we observe CDR3 sequences longer than <code>40</code> amino acids at
a rate of about <code>1/10,000</code> cells, and these are typically part of a <i>third</i> chain
for a given cell, suggesting that such chains are likely nonproductive.  However it may be that
long CDR3 sequences occur as part of productive chains in other species.
</li>
<li>
Add a chain variable <code>cdr2_aa_ref</code> the gives the value of the CDR2 amino acids
for the universal reference sequence, and similarly for other cases, see
<a href="../../pages/auto/help.cvars.html"><code>enclone help cvars</code></a>.
</li>
</ul>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>11/20/20</b>: 
<ul>
<li>
Generalize the <code>FCELL</code> argument to allow arbitrary expressions.  Please see
<a href="../../pages/auto/help.special.html"><code>enclone help special</code></a>.
<span style="color:red">(This is a breaking change.)</span>
</li>
</ul>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>11/19/20</b>: 
<ul>
<li>Add lead variable <code>count_<regex></code> that counts the number of occurrences of a given
pattern in the V..J sequences for all chains.</li>
</ul>
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>11/4/20</b>: 
<ul>
<li>Add <code>SUBSET_JSON</code> option, which allows construction of a small json file
associated to specified clonotypes, which can be sent to the enclone developers.</li>
<li>The <code>NOPRINT</code> option no longer applies to parseable output generated using 
<code>POUT=stdout</code> or <code>POUT=stdouth</code>.  This change facilitates piping of parseable 
output to other programs.</li>
</ul>
</p>

<hr>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>10/20/20</b>: If a variable is of the form <code>abbr:name</code>, and <code>POUT</code> is
set to a file name, then the field label that is shown is now <code>abbr</code>, rather than
<code>abbr:name</code>.  This is really a bug fix.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>10/9/20</b>: Add parseable variables for the start and frame of the D region.
</p>

<hr>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>9/17/20</b>: Clarify documentation for the case where <code>GEX=(antibody data)</code> and fix 
bug in handling of <code>n_gex</code> for that.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>9/16/20</b>: 
<ul>
<li>
<span style="color:red">
Test for defective reference sequences and exit if they're found (allowing for override).
</span>
There is also a new argument <code>BUILT_IN</code> to force recomputation using the built-in
reference (for human or mouse).
<li>We turned off part of the graph filtering (which can be turned off completely with the
argument <code>NGRAPH_FILTER</code>).  We did this because in some cases this code was deleting
significant numbers of real cells.
</li>
<li>
Tweak feature computations.  The page
<a href="../../pages/auto/features.html">this page</a> has been updated accordingly.  The
results for tests of species described on that page did not change, but there were changes to
other species, including a few changes for dog and horse.
</li>
</ul>
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>9/4/20</b>: Add the capability of seeing which default filters would have been applied, if
the filters are turned off.  See the lead variable "filter" at 
<a href="../../pages/auto/help.lvars.html"><code>enclone help lvars</code></a>.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>9/2/20</b>:
<ul>
<li>Add options <code>DIFF_STYLE=C1</code> and <code>DIFF_STYLE=C2</code> that allow alternate
labeling for the diff row in a clonotype table, which by default shows x's and dots.</li>
<li>Fix bugs in the diff row printing: sometimes it was not shown, and it was not correctly
handled when <code>COLOR=property</code> was specified.</li>
</ul>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>9/1/20</b>:
<ul>
<li>Add parseable field <code>vj_seq_nl</code>, the DNA sequence of V..J, but starting after the
leader.</li>
<li>Add parseable field <code>vj_aa_nl</code>, the amino acid sequence of V..J, but starting after
the leader.</li>
<li>Add chain variable <code>aa%</code> = amino acid percent identity with donor reference, 
outside junction region.</li>
<li>Add chain variable <code>dna%</code> = nucleotide percent identity with donor reference, 
outside junction region.</li>
<li>Add summary of barcode fate if SUMMARY option is used.</li>
</ul>
</p

<hr>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>8/31/20</b>: add FWR4 support
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>8/28/20</b>: add new filters <code>CONST_IGH</code> and <code>CONST_IGKL</code> that allow
filtering of cells by isotype.  Please see
<a href="../../pages/auto/help.special.html"><code>enclone help special</code></a>.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>8/27/20</b></b>: 
<ol>
<li>
Fix bugs in reporting of insertions in the notes field.
</li>
<li>
Now CDR1, CDR2, FWR1, FWR2 and FWR3 can be can be computed and displayed.  Please see
<a href="../../pages/auto/help.amino.html"><code>enclone help amino</code></a> 
and <a href="../../pages/auto/help.cvars.html"><code>enclone help cvars</code></a> 
for how to specify these, and 
<a href="../../pages/auto/features.html">this page</a> for how we calculate these features.
</li>
</ol>
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>8/24/20</b></b>: 
<ol>
<li>Add lead variable <code>dref_aa</code>, the number of amino acid differences
with the donor reference.</li>
<li>Add estimated doublet rate to summary stats.</li>
<li>We are deprecating the lead variables <code>sec</code> and <code>mem</code> that measure
UMIs supporting secreted versus membrane-bound antibody transcripts.  Because the exon boundaries
that determine this are far from the 5' end, there is rarely enough signal to be informative.
We removed the documentation.</li>
</ol>
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>8/20/20</b></b>: 
<ol>
<li> Allow "Levenshtein distance patterns" for CDR3 filter specification.  For more information,
type or click here:
<a href="../../pages/auto/help.filter.html"><code>enclone help filter</code></a>.</li>
<li> Change what the weak onesies filter does: unless <code>NWEAK_ONESIES</code> is used, now,
rather than delete certain dubious onesie clonotypes, such clonotypes are disintegrated into
single cell clonotypes.</li>
<li> Change what the improper filter does: unless <code>NIMPROPER</code> is used, now,
clonotypes having more than one chain but all of the same type are deleted.</li>
<li> Add a new filter that prevents merges of onesie clonotypes into other clonotypes; this
filter may be turned off with a new argument <code>NMERGE_ONESIES</code>.  This behavior is
mostly (but not completely) a splitting out of previous functionality so as to simplify and
clarify.</li>
</ol>
Importantly, the last three changes have the effect of <i>increasing</i> the number of cells in
clonotypes; the extra cells are mostly in onesie clonotypes.
Precise definitions may be found by typing (or clicking here): 
<a href="../../pages/auto/help.special.html"><code>enclone help special</code></a>.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>8/17/20</b></b>: deprecate <code>KEEP_IMPROPER</code>; the new name is 
<code>NIMPROPER</code>.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>8/10/20</b></b>: tweak the definition of the weak onesies filter so that it does not
delete single-cell clonotypes.  Superceded by subsequent change on 8/18/20.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>8/7/20</b></b>: add the ability to analyze alternate splicing using GEX data to characterize
UMIs as secreted or membrane, and display this information using lvars "sec" and "mem".   See
<code>enclone help lvars</code> for limitations.  This is highly experimental.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>8/5/20</b></b>:
<ol>
<li>A reference file is now required as part of the Cell Ranger outs directory,
if Cell Ranger version 4.0 or greater was used.</li>
<li>Deprecate <code>MAX_SCORE</code> and replace by <code>MAX_LOG_SCORE</code>, the
    base 10 logarithm of it.</li>
<li>Add cvar <code>cdr3_len</code></li>
<li>Add and document knobs that allow nearly all clonotype join filtering to be turned off.  Please
    see "enclone help how" and the end of "enclone help faq".</li>
</ol>
</p>

<hr>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>6/24/20</b>:
<ol>
<li>Add <a href="../../pages/auto/innate.html">support for iNKT and MAIT cells</a>.</li>
<li>Make <span style="color:rgb(120,123,175);font-weight:900">enclone</span> faster.  This is most noticeable in cases where many GEX datasets are
    provided as input.</li>
</ol>
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>6/19/20</b>: 
<ol>
<li>We now use a data hierarchy of donor (top), origin, dataset (bottom), where an 
origin is a set of 1 or more datasets from the same source (tube of cells, tissue, timepoint, etc.).
<span style="color:red">(This breaks previous invocations of <code>META</code>.)</span>
</li>
<li> Improve the alternate (faster) internal storage structure for the GEX matrix
created using the option <code>NH5</code> as described using the command 
<a href="../../pages/auto/help.input.html"><code>enclone help input</code></a>.  
This will speed things up, particularly for the case where several datasets are combined.  If 
you have already used the <code>NH5</code> option for a given dataset, then the next time you
run <span style="color:rgb(120,123,175);font-weight:900">enclone</span> on it, the file will be automatically rewritten.  This would also apply to some
datasets obtained as part of the <code>large</code> download.
</li>
</ol>
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>6/17/20</b>:
<ol>
<li>Now <code>TREE=const</code> can be used to show a tree with heavy chain constant region
    names attached to the leaves. </li>
<li><code>SEG</code> and <code>SEGN</code> are now cumulative, so that multiple instances may
    be used to progressively filter.</li>
<li>The clonotype joining heuristic parameters <code>MAX_SCORE</code> and
    <code>MAX_CDR3_DIFFS</code> are now accessible.</li>
</ol>
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>6/10/20</b>:
<ol>
<li>Add a complex of features for generating phylogenetic trees from clonotypes, see
    <a href="../../pages/auto/tree.html">here</a>.</li>
<li>New "single button" installation procedure.</li>
<li>Change the default value for <code>PRE</code> 
    to <code>~/enclone/datasets,~/enclone/datasets2</code>.</li>
<li>Add argument <code>NALL</code> to turn off all filters.</li>
<li>Add new lead variables <code>nd&lt;k></code> that display the number of cells in the
    top datasets for a given clonotype.</li>
<li>Add new lead variable <code>dref</code> that shows the distance of V..J from the reference
    outside the region of recombination.
<li>Add argument <code>COMPLETE</code> to remove exact subclonotypes that do not have all 
    chains.</li>
<li>Test for consistency between VDJ and GEX barcodes, and exit if this is not the case.</li>
<li>Add option <code>COLOR=property</code> to color amino acids by their properties.</li>
<li>Add option <code>FCELL</code> to allow filtering by cells.</li>
</ol>
</p>

<hr>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>5/29/20</b>:
Add new "UMI ratio" filter that further reduces noise in certain cases.  This can be turned
off using the argument <code>NUMI_RATIO</code>.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>5/22/20</b>:
Add major new "UMI" filter that greatly reduces noise in certain cases.  This can be turned
off using the argument <code>NUMI</code>.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>5/12/20</b>:
Add <code>PLOT_BY_ISOTYPE</code> to generate honeycomb plots colored by isotype.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>5/1/20</b>: 
Change the definition of the fields "edit" and "comp" to be based on
alignment from the beginning of the CDR3 up to the end of the J, rather than stopping at
the end of the CDR3.  The intention is to capture the full region of recombination, which
may not have been done before.
</p>

<hr>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

<p>
<b>4/30/20</b>: 
First release.
</p>

<! â€“â€“ â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“â–“ -->

</body>
</html>
