<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" 
"https://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<!--  -->
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="application/xml+xhtml; charset=UTF-8"/>
<title>enclone history</title>
<link rel="stylesheet" type="text/css" href="../enclone_css_v2.css">
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-58278925-3"></script>
<script>
window.dataLayer = window.dataLayer || [];
function gtag(){{dataLayer.push(arguments);}}
gtag('js', new Date());
gtag('config', 'UA-58278925-3');
</script>

        
        </head>

        <! â€“â€“

        ðŸ’© ðŸ’© ðŸ’© ðŸ”´ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”´ ðŸ’© ðŸ’© ðŸ’©

        PUT DOWN YOUR HAMMER.
        THIS IS AN AUTO-GENERATED FILE.  PLEASE DO NOT EDIT IT.
        THANK YOU FOR YOUR COOPERATION,

        SINCERELY,
        THE BENEVOLENT OVERLORDS

        ðŸ’© ðŸ’© ðŸ’© ðŸ”´ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”¨ ðŸ”´ ðŸ’© ðŸ’© ðŸ’©

        â€“â€“>

<body>

<br>
<img src="../../img/enclone_banner.png" alt="enclone banner" title="enclone banner" width=100% />

<h1>History</h1>

<p>This page provides a selective history of what was changed in <span style="color:rgb(120,123,175);font-weight:900">enclone</span> and when.</p>

<p>
We show changes that affect users like new features, changes to results, and the like.  This
log starts with with initial public availability.  The complete history may be seen by cloning the 
<span style="color:rgb(120,123,175);font-weight:900">enclone</span> repo and typing <code>git log</code>.
</p>

<p><span style="color:red">Breaking changes are shown in red.</span></p>

<p style="border: 2px; border-style: solid; border-color: black; background-color: #EBF4FC;
padding: 8px; width: 900px; font-size: 110%">
Please be aware that our workflow when we make changes is to automatically update the GitHub
site, including all the website pages and this page too.  This happens <i>in advance</i> of 
actually making a release (which might follow in a couple days).  This means that the website
may describe features that are not yet available in a release (although they will be in the
source code that's available).  We apologize for this asynchrony!  Note however that the 
command-line help that comes with your copy of <span style="color:rgb(120,123,175);font-weight:900">enclone</span> will always match its behavior.</p>

<hr>

<p>
<b>h/8/31/20</b>: add FWR4 support
</p>

<p>
<b>8/28/20</b>: add new filters <code>CONST_IGH</code> and <code>CONST_IGKL</code> that allow
filtering of cells by isotype.  Please see
<a href="../../pages/auto/help.special.html"><code>enclone help special</code></a>.
</p>

<p>
<b>8/27/20</b></b>: 
<ol>
<li>
Fix bugs in reporting of insertions in the notes field.
</li>
<li>
Now CDR1, CDR2, FWR1, FWR2 and FWR3 can be can be computed and displayed.  Please see
<a href="../../pages/auto/help.amino.html"><code>enclone help amino</code></a> 
and <a href="../../pages/auto/help.cvars.html"><code>enclone help cvars</code></a> 
for how to specify these, and 
<a href="../../pages/auto/features.html">this page</a> for how we calculate these features.
</li>
</ol>
</p>

<p>
<b>8/24/20</b></b>: 
<ol>
<li>Add lead variable <code>dref_aa</code>, the number of amino acid differences
with the donor reference.</li>
<li>Add estimated doublet rate to summary stats.</li>
<li>We are deprecating the lead variables <code>sec</code> and <code>mem</code> that measure
UMIs supporting secreted versus membrane-bound antibody transcripts.  Because the exon boundaries
that determine this are far from the 5' end, there is rarely enough signal to be informative.
We removed the documentation.</li>
</ol>
</p>

<p>
<b>8/20/20</b></b>: 
<ol>
<li> Allow "Levenshtein distance patterns" for CDR3 filter specification.  For more information,
type or click here:
<a href="../../pages/auto/help.filter.html"><code>enclone help filter</code></a>.</li>
<li> Change what the weak onesies filter does: unless <code>NWEAK_ONESIES</code> is used, now,
rather than delete certain dubious onesie clonotypes, such clonotypes are disintegrated into
single cell clonotypes.</li>
<li> Change what the improper filter does: unless <code>NIMPROPER</code> is used, now,
clonotypes having more than one chain but all of the same type are deleted.</li>
<li> Add a new filter that prevents merges of onesie clonotypes into other clonotypes; this
filter may be turned off with a new argument <code>NMERGE_ONESIES</code>.  This behavior is
mostly (but not completely) a splitting out of previous functionality so as to simplify and
clarify.</li>
</ol>
Importantly, the last three changes have the effect of <i>increasing</i> the number of cells in
clonotypes; the extra cells are mostly in onesie clonotypes.
Precise definitions may be found by typing (or clicking here): 
<a href="../../pages/auto/help.special.html"><code>enclone help special</code></a>.
</p>

<p>
<b>8/17/20</b></b>: deprecate <code>KEEP_IMPROPER</code>; the new name is 
<code>NIMPROPER</code>.
</p>

<p>
<b>8/10/20</b></b>: tweak the definition of the weak onesies filter so that it does not
delete single-cell clonotypes.  Superceded by subsequent change on 8/18/20.
</p>

<p>
<b>8/7/20</b></b>: add the ability to analyze alternate splicing using GEX data to characterize
UMIs as secreted or membrane, and display this information using lvars "sec" and "mem".   See
<code>enclone help lvars</code> for limitations.  This is highly experimental.
</p>

<p>
<b>8/5/20</b></b>:
<ol>
<li>A reference file is now required as part of the Cell Ranger outs directory,
if Cell Ranger version 4.0 or greater was used.</li>
<li>Deprecate <code>MAX_SCORE</code> and replace by <code>MAX_LOG_SCORE</code>, the
    base 10 logarithm of it.</li>
<li>Add cvar <code>cdr3_len</code></li>
<li>Add and document knobs that allow nearly all clonotype join filtering to be turned off.  Please
    see \"enclone help how\" and the end of \"enclone help faq\".</li>
</ol>
</p>

<p>
<b>6/24/20</b>:
<ol>
<li>Add <a href="../../pages/auto/innate.html">support for iNKT and MAIT cells</a>.</li>
<li>Make <span style="color:rgb(120,123,175);font-weight:900">enclone</span> faster.  This is most noticeable in cases where many GEX datasets are
    provided as input.</li>
</ol>
</p>

<p>
<b>6/19/20</b>: 
<ol>
<li>We now use a data hierarchy of donor (top), origin, dataset (bottom), where an 
origin is a set of 1 or more datasets from the same source (tube of cells, tissue, timepoint, etc.).
<span style="color:red">(This breaks previous invocations of <code>META</code>.)</span>
</li>
<li> Improve the alternate (faster) internal storage structure for the GEX matrix
created using the option <code>NH5</code> as described using the command 
<a href="../../pages/auto/help.input.html"><code>enclone help input</code></a>.  
This will speed things up, particularly for the case where several datasets are combined.  If 
you have already used the <code>NH5</code> option for a given dataset, then the next time you
run <span style="color:rgb(120,123,175);font-weight:900">enclone</span> on it, the file will be automatically rewritten.  This would also apply to some
datasets obtained as part of the <code>large</code> download.
</li>
</ol>
</p>

<p>
<b>6/17/20</b>:
<ol>
<li>Now <code>TREE=const</code> can be used to show a tree with heavy chain constant region
    names attached to the leaves. </li>
<li><code>SEG</code> and <code>SEGN</code> are now cumulative, so that multiple instances may
    be used to progressively filter.</li>
<li>The clonotype joining heuristic parameters <code>MAX_SCORE</code> and
    <code>MAX_CDR3_DIFFS</code> are now accessible.</li>
</ol>
</p>

<p>
<b>6/10/20</b>:
<ol>
<li>Add a complex of features for generating phylogenetic trees from clonotypes, see
    <a href="../../pages/auto/tree.html">here</a>.</li>
<li>New "single button" installation procedure.</li>
<li>Change the default value for <code>PRE</code> 
    to <code>~/enclone/datasets,~/enclone/datasets2</code>.</li>
<li>Add argument <code>NALL</code> to turn off all filters.</li>
<li>Add new lead variables <code>nd&lt;k></code> that display the number of cells in the
    top datasets for a given clonotype.</li>
<li>Add new lead variable <code>dref</code> that shows the distance of V..J from the reference
    outside the region of recombination.
<li>Add argument <code>COMPLETE</code> to remove exact subclonotypes that do not have all 
    chains.</li>
<li>Test for consistency between VDJ and GEX barcodes, and exit if this is not the case.</li>
<li>Add option <code>COLOR=property</code> to color amino acids by their properties.</li>
<li>Add option <code>FCELL</code> to allow filtering by cells.</li>
</ol>
</p>

<p>
<b>5/29/20</b>:
Add new "UMI ratio" filter that further reduces noise in certain cases.  This can be turned
off using the argument <code>NUMI_RATIO</code>.
</p>

<p>
<b>5/22/20</b>:
Add major new "UMI" filter that greatly reduces noise in certain cases.  This can be turned
off using the argument <code>NUMI</code>.
</p>

<p>
<b>5/12/20</b>:
Add <code>PLOT_BY_ISOTYPE</code> to generate honeycomb plots colored by isotype.
</p>

<p>
<b>5/1/20</b>: 
Change the definition of the fields "edit" and "comp" to be based on
alignment from the beginning of the CDR3 up to the end of the J, rather than stopping at
the end of the CDR3.  The intention is to capture the full region of recombination, which
may not have been done before.
</p>

<p>
<b>4/30/20</b>: 
First release.
</p>

</body>
</html>
